--- a/configure
+++ b/configure
@@ -26319,9 +26319,6 @@ ac_compiler_gnu=$ac_cv_c_compiler_gnu
  ex_check_save_LIBS=${LIBS}
  ex_check_save_CPPFLAGS=${CPPFLAGS}
  ex_check_save_LDFLAGS=${LDFLAGS}
- if test "x/usr/include/libart-2.0" != "x"; then
-   CPPFLAGS="$CPPFLAGS -I/usr/include/libart-2.0"
- fi
   { echo "$as_me:$LINENO: checking for art_vpath_add_point in -lart_lgpl_2" >&5
 echo $ECHO_N "checking for art_vpath_add_point in -lart_lgpl_2... $ECHO_C" >&6; }
 if test "${ac_cv_lib_art_lgpl_2_art_vpath_add_point+set}" = set; then
@@ -27915,9 +27912,6 @@ ac_compiler_gnu=$ac_cv_c_compiler_gnu
  ex_check_save_LIBS=${LIBS}
  ex_check_save_CPPFLAGS=${CPPFLAGS}
  ex_check_save_LDFLAGS=${LDFLAGS}
- if test "x/usr/include/freetype2" != "x"; then
-   CPPFLAGS="$CPPFLAGS -I/usr/include/freetype2"
- fi
   { echo "$as_me:$LINENO: checking for FT_Init_FreeType in -lfreetype" >&5
 echo $ECHO_N "checking for FT_Init_FreeType in -lfreetype... $ECHO_C" >&6; }
 if test "${ac_cv_lib_freetype_FT_Init_FreeType+set}" = set; then
--- a/acinclude.m4
+++ b/acinclude.m4
@@ -20,9 +20,6 @@ AC_DEFUN([EX_CHECK_ALL],
  ex_check_save_LIBS=${LIBS}
  ex_check_save_CPPFLAGS=${CPPFLAGS}
  ex_check_save_LDFLAGS=${LDFLAGS}
- if test "x$7" != "x"; then
-   CPPFLAGS="$CPPFLAGS -I$7"
- fi
  dnl try compiling naked first
  AC_CHECK_LIB($1,$2, [
     AC_CHECK_HEADER($3,[LIBS="-l$1 ${LIBS}";EX_CHECK_STATE=YES],[])],[])
