--- a/Makefile.am
+++ b/Makefile.am
@@ -1,7 +1,7 @@
 ## Process this file with automake to produce Makefile.in
 
-SUBDIRS = src examples misc
-DIST_SUBDIRS = $(SUBDIRS) doc debian wince
+SUBDIRS = src
+DIST_SUBDIRS = $(SUBDIRS) debian wince
 
 EXTRA_DIST = libdvbpsi.spec bootstrap
 
--- a/Makefile.in
+++ b/Makefile.in
@@ -179,8 +179,8 @@ target_alias = @target_alias@
 target_cpu = @target_cpu@
 target_os = @target_os@
 target_vendor = @target_vendor@
-SUBDIRS = src examples misc
-DIST_SUBDIRS = $(SUBDIRS) doc debian wince
+SUBDIRS = src
+DIST_SUBDIRS = $(SUBDIRS) debian wince
 EXTRA_DIST = libdvbpsi.spec bootstrap
 AUTOMAKE_OPTIONS = foreign dist-bzip2 subdir-objects
 all: config.h
