--- a/Makefile.in
+++ b/Makefile.in
@@ -110,7 +110,7 @@ RECURSIVE_CLEAN_TARGETS = mostlyclean-re
 ETAGS = etags
 CTAGS = ctags
 DIST_SUBDIRS = support genfft kernel simd dft rdft reodft api threads \
-	cell libbench2 . tests doc tools m4
+	cell libbench2 . doc tools m4
 DISTFILES = $(DIST_COMMON) $(DIST_SOURCES) $(TEXINFOS) $(EXTRA_DIST)
 distdir = $(PACKAGE)-$(VERSION)
 top_distdir = $(distdir)
@@ -278,7 +278,7 @@ lib_LTLIBRARIES = libfftw3@PREC_SUFFIX@.
 @MAINTAINER_MODE_TRUE@GENFFT = genfft
 ACLOCAL_AMFLAGS = -I m4
 SUBDIRS = support $(GENFFT) kernel simd dft rdft reodft api threads \
-cell libbench2 . tests doc tools m4
+cell libbench2 . m4
 
 EXTRA_DIST = COPYRIGHT bootstrap.sh CONVENTIONS fftw.pc.in README.Cell
 @HAVE_SIMD_TRUE@SIMD_LIBS = \
