diff -ur tctool.old/src/Makefile tctool/src/Makefile
--- tctool.old/src/Makefile	2007-01-03 02:32:24.000000000 +0100
+++ tctool/src/Makefile	2007-09-18 19:15:03.068155702 +0200
@@ -31,16 +31,16 @@
 EXAMPLE  := example
 EXOBJ    := if.o interface.o etherStream.o etherStreambuf.o tcStream.o tcStreambuf.o tcRtsDump.o tcRtsDumpbuf.o
 
-CXX        := g++
-CXXFLAGS   := -march=i386 -O2
-LDFLAGS    := --strip-unneeded
-LIBS       := 
-LDSTATIC   := -L. -static-libgcc
-STATICLIBS := stdc++ m
-COUNTRY    := "C"
-LANG       := "C"
-LANGUAGE   := "C"
-LC_ALL     := "C"
+#CXX        := g++
+#CXXFLAGS   := -march=i386 -O2
+#LDFLAGS    := --strip-unneeded
+#LIBS       := 
+#LDSTATIC   := -L. -static-libgcc
+#STATICLIBS := stdc++ m
+#COUNTRY    := "C"
+#LANG       := "C"
+#LANGUAGE   := "C"
+#LC_ALL     := "C"
 
 default: install
 
