diff -urN mgen.old/protolib/unix/Makefile.linux mgen/protolib/unix/Makefile.linux
--- mgen.old/protolib/unix/Makefile.linux      2008-03-07 22:23:06.000000000 +0100
+++ mgen/protolib/unix/Makefile.linux  2009-06-06 20:40:29.000000000 +0200
@@ -39,7 +39,7 @@
 # (We export these for other Makefiles as needed)
 #
 
-SYSTEM_HAVES = -DLINUX -DHAVE_IPV6 -DHAVE_GETLOGIN -D_FILE_OFFSET_BITS=64 -DHAVE_LOCKF \
+SYSTEM_HAVES = -DLINUX -DHAVE_GETLOGIN -D_FILE_OFFSET_BITS=64 -DHAVE_LOCKF \
 -DHAVE_OLD_SIGNALHANDLER -DHAVE_DIRFD -DHAVE_ASSERT -DNO_SCM_RIGHTS -DHAVE_SCHED
 
 # (TBD) Move ProtoRouteMgr to ProtokitEx ??
diff -urN mgen.old/unix/Makefile.linux mgen/unix/Makefile.linux
--- mgen.old/unix/Makefile.linux       2007-06-22 21:16:20.000000000 +0200
+++ mgen/unix/Makefile.linux   2009-06-06 20:40:10.000000000 +0200
@@ -35,7 +35,7 @@
 # (We export these for other Makefiles as needed)
 #
 
-SYSTEM_HAVES = -DLINUX -DHAVE_PSELECT -DHAVE_SCHED -DHAVE_IPV6 -DHAVE_GETLOGIN -DHAVE_LOCKF -DHAVE_DIRFD \
+SYSTEM_HAVES = -DLINUX -DHAVE_PSELECT -DHAVE_SCHED -DHAVE_GETLOGIN -DHAVE_LOCKF -DHAVE_DIRFD \
 -DHAVE_ASSERT $(NETSEC)
 
 SYSTEM = linux
