--- a/configure
+++ b/configure
@@ -5362,19 +5362,6 @@ _ACEOF
   FRAMER="generic"
 fi
 
-if test $FRAMER = "linux"; then
-   echo "$as_me:$LINENO: checking for procfs support" >&5
-echo $ECHO_N "checking for procfs support... $ECHO_C" >&6
-   if test -r /proc/net/dev; then
-      echo "$as_me:$LINENO: result: okay" >&5
-echo "${ECHO_T}okay" >&6
-   else
-      { { echo "$as_me:$LINENO: error: You must have ProcFS enabled in your kernel!" >&5
-echo "$as_me: error: You must have ProcFS enabled in your kernel!" >&2;}
-   { (exit 1); exit 1; }; };
-   fi
-fi
-
 CFLAGS="$GGDBFLAGS $CFLAGS"
 
 touch Makefile.in
