Index: xsupplicant-1.2.8/src/cardif/linux/cardif_atmel_driver.c
===================================================================
--- xsupplicant-1.2.8.orig/src/cardif/linux/cardif_atmel_driver.c	2006-10-08 05:42:53.000000000 +0200
+++ xsupplicant-1.2.8/src/cardif/linux/cardif_atmel_driver.c	2009-01-11 02:59:46.998587770 +0100
@@ -103,7 +103,6 @@
 #include <sys/ioctl.h>
 #include <linux/types.h>
 #include <linux/socket.h>
-#include <linux/compiler.h>
 #include <iwlib.h>
 #include <stdlib.h>
 #include <string.h>
Index: xsupplicant-1.2.8/src/cardif/linux/cardif_linux.c
===================================================================
--- xsupplicant-1.2.8.orig/src/cardif/linux/cardif_linux.c	2006-10-08 05:42:53.000000000 +0200
+++ xsupplicant-1.2.8/src/cardif/linux/cardif_linux.c	2009-01-11 03:00:04.496300396 +0100
@@ -174,7 +174,6 @@
 #include <netinet/in.h>
 #include <linux/types.h>
 #include <linux/socket.h>
-#include <linux/compiler.h>
 #include <iwlib.h>
 #include <linux/if_packet.h>
 #include <stdio.h>
Index: xsupplicant-1.2.8/src/cardif/linux/cardif_linux_rtnetlink.c
===================================================================
--- xsupplicant-1.2.8.orig/src/cardif/linux/cardif_linux_rtnetlink.c	2006-10-08 05:42:53.000000000 +0200
+++ xsupplicant-1.2.8/src/cardif/linux/cardif_linux_rtnetlink.c	2009-01-11 03:00:14.997328194 +0100
@@ -203,7 +203,6 @@
 
 #include <unistd.h>
 #include <sys/socket.h>
-#include <linux/compiler.h>
 #include <iwlib.h>
 #include <sys/ioctl.h>
 #include "netlink.h"
Index: xsupplicant-1.2.8/src/cardif/linux/cardif_linux_wext.c
===================================================================
--- xsupplicant-1.2.8.orig/src/cardif/linux/cardif_linux_wext.c	2009-01-11 02:58:02.000000000 +0100
+++ xsupplicant-1.2.8/src/cardif/linux/cardif_linux_wext.c	2009-01-11 03:00:26.230427636 +0100
@@ -203,7 +203,6 @@
 #include <netinet/in.h>
 #include <linux/types.h>
 #include <linux/socket.h>
-#include <linux/compiler.h>
 #include <linux/wireless.h>
 #include <iwlib.h>
 #include <linux/if_packet.h>
