diff -urN ssldump-0.9b3/base/pcap-snoop.c ssldump-0.9b3.new/base/pcap-snoop.c
--- ssldump-0.9b3/base/pcap-snoop.c	2002-09-09 23:02:58.000000000 +0200
+++ ssldump-0.9b3.new/base/pcap-snoop.c	2008-12-14 18:43:48.000000000 +0100
@@ -49,7 +49,7 @@
 
 #include <pcap.h>
 #include <unistd.h>
-#include <net/bpf.h>
+#include <pcap-bpf.h>
 #ifndef _WIN32
 #include <sys/param.h>
 #endif
diff -urN ssldump-0.9b3/ssl/ssldecode.c ssldump-0.9b3.new/ssl/ssldecode.c
--- ssldump-0.9b3/ssl/ssldecode.c	2002-08-17 03:33:17.000000000 +0200
+++ ssldump-0.9b3.new/ssl/ssldecode.c	2008-12-14 18:44:11.000000000 +0100
@@ -52,6 +52,7 @@
 #include <openssl/hmac.h>
 #include <openssl/evp.h>
 #include <openssl/x509v3.h>
+#include <openssl/md5.h>
 #endif
 #include "ssldecode.h"
 #include "ssl_rec.h"
