diff --git a/common/os/posix/IfConfigLinux.cc b/common/os/posix/IfConfigLinux.cc
index d846259..ca91bf2 100644
--- a/common/os/posix/IfConfigLinux.cc
+++ b/common/os/posix/IfConfigLinux.cc
@@ -380,7 +380,7 @@ static std::list<IfEntry> NetlinkGetInterfaces(void)
     // everyone else does and just allocate a  "big enough' buffer.  We choose
     // 64K which should handle about 150 interfaces.
     //
-    const uint32_t BUFSIZE = 65536;
+    const uint32_t BUFSIZE = 262144;
     char* buffer = new char[BUFSIZE];
     uint32_t len;
