--- a/vtysh/vtysh.c
+++ b/vtysh/vtysh.c
@@ -268,7 +268,7 @@ vtysh_pager_init (void)
   if (pager_defined)
     vtysh_pager_name = strdup (pager_defined);
   else
-    vtysh_pager_name = strdup ("more");
+    vtysh_pager_name = strdup ("cat");
 }
 
 /* Command execution over the vty interface. */
@@ -1884,7 +1884,7 @@ DEFUN (vtysh_terminal_length,
 {
   int lines;
   char *endptr = NULL;
-  char default_pager[10];
+  char default_pager[12];
 
   lines = strtol (argv[0], &endptr, 10);
   if (lines < 0 || lines > 512 || *endptr != '\0')
@@ -1901,7 +1901,7 @@ DEFUN (vtysh_terminal_length,
 
   if (lines != 0)
     {
-      snprintf(default_pager, 10, "more -%i", lines);
+      snprintf(default_pager, 12, "head -n %i", lines);
       vtysh_pager_name = strdup (default_pager);
     }
 
@@ -2211,9 +2211,9 @@ void
 vtysh_readline_init (void)
 {
   /* readline related settings. */
-  rl_bind_key ('?', (Function *) vtysh_rl_describe);
+  rl_bind_key ('?', (rl_command_func_t *) vtysh_rl_describe);
   rl_completion_entry_function = vtysh_completion_entry_function;
-  rl_attempted_completion_function = (CPPFunction *)new_completion;
+  rl_attempted_completion_function = (rl_completion_func_t *) new_completion;
   /* do not append space after completion. It will be appended
    * in new_completion() function explicitly. */
   rl_completion_append_character = '\0';
