--- a/libntfs/Makefile.am
+++ b/libntfs/Makefile.am
@@ -38,8 +38,8 @@ libntfs_la_CFLAGS = $(LIBNTFS_CFLAGS) \
 	-DLTVERSION_LIBNTFS=\"$(LTVERSION_LIBNTFS)\"
 
 if ENABLE_CRYPTO
-libntfs_la_LDFLAGS += `libgnutls-config --libs` `pkg-config --libs libconfig`
-libntfs_la_CFLAGS += `libgnutls-config --cflags` `pkg-config --cflags libconfig`
+libntfs_la_LDFLAGS += $(GNUTLS_LIBS) $(libconfig_LIBS)
+libntfs_la_CFLAGS += $(GNUTLS_CFLAGS) $(libconfig_CFLAGS)
 endif
 
 libntfs_la_SOURCES =	\
--- a/libntfs/Makefile.in
+++ b/libntfs/Makefile.in
@@ -61,8 +61,8 @@ POST_UNINSTALL = :
 build_triplet = @build@
 host_triplet = @host@
 target_triplet = @target@
-@ENABLE_CRYPTO_TRUE@am__append_1 = `libgnutls-config --libs` `pkg-config --libs libconfig`
-@ENABLE_CRYPTO_TRUE@am__append_2 = `libgnutls-config --cflags` `pkg-config --cflags libconfig`
+@ENABLE_CRYPTO_TRUE@am__append_1 = $(GNUTLS_LIBS) $(libconfig_LIBS)
+@ENABLE_CRYPTO_TRUE@am__append_2 = $(GNUTLS_CFLAGS) $(libconfig_CFLAGS)
 subdir = libntfs
 DIST_COMMON = $(srcdir)/Makefile.am $(srcdir)/Makefile.in \
 	$(srcdir)/libntfs-gnomevfs.8.in $(srcdir)/libntfs.8.in \
--- a/ntfsprogs/Makefile.am
+++ b/ntfsprogs/Makefile.am
@@ -125,8 +125,8 @@ ntfsdump_logfile_LDFLAGS= $(AM_LFLAGS)
 if ENABLE_CRYPTO
 ntfsdecrypt_SOURCES	= ntfsdecrypt.c utils.c utils.h
 ntfsdecrypt_LDADD	= $(AM_LIBS)
-ntfsdecrypt_LDFLAGS	= $(AM_LFLAGS) `libgnutls-config --libs`
-ntfsdecrypt_CFLAGS	= `libgnutls-config --cflags`
+ntfsdecrypt_LDFLAGS	= $(AM_LFLAGS) $(GNUTLS_LIBS)
+ntfsdecrypt_CFLAGS	= $(GNUTLS_CFLAGS)
 endif
 
 # Extra targets
--- a/ntfsprogs/Makefile.in
+++ b/ntfsprogs/Makefile.in
@@ -449,8 +449,8 @@ ntfsdump_logfile_LDADD = $(AM_LIBS)
 ntfsdump_logfile_LDFLAGS = $(AM_LFLAGS)
 @ENABLE_CRYPTO_TRUE@ntfsdecrypt_SOURCES = ntfsdecrypt.c utils.c utils.h
 @ENABLE_CRYPTO_TRUE@ntfsdecrypt_LDADD = $(AM_LIBS)
-@ENABLE_CRYPTO_TRUE@ntfsdecrypt_LDFLAGS = $(AM_LFLAGS) `libgnutls-config --libs`
-@ENABLE_CRYPTO_TRUE@ntfsdecrypt_CFLAGS = `libgnutls-config --cflags`
+@ENABLE_CRYPTO_TRUE@ntfsdecrypt_LDFLAGS = $(AM_LFLAGS) $(GNUTLS_LIBS)
+@ENABLE_CRYPTO_TRUE@ntfsdecrypt_CFLAGS = $(GNUTLS_CFLAGS)
 all: all-am
 
 .SUFFIXES:
