--- a/FAQ
+++ b/FAQ
@@ -99,7 +99,7 @@ TABLE OF CONTENTS
    I don't understand the new access control stuff - what does it mean?
    How do I configure SNMPv3 users?
    The 'createUser' line disappears when I start the agent.  Why?
-   What's the difference between /var/net-snmp and /usr/local/share/snmp?
+   What's the difference between /var/lib/snmp and /usr/share/snmp?
    My new agent is ignoring the old snmpd.conf file. Why?
    Why am I getting "Connection refused"?
    Why can't I see values in the UCDavis 'extensible' or 'disk' trees?
@@ -643,7 +643,7 @@ How do I add a MIB to the tools?
   There are two steps required to add a new MIB file to the tools.
   Firstly, copy the MIB file into the appropiate location:
 
-	cp MY-MIB.txt /usr/local/share/snmp/mibs
+	cp MY-MIB.txt /usr/share/snmp/mibs
             (which makes it available to everyone on the system)
 
           or
@@ -1177,7 +1177,7 @@ My traphandler script doesn't work when 
   If this is the case, then you can specify this interpreter
   explicitly as part of the trap handle directive:
 
-	traphandle default /usr/bin/perl /usr/local/bin/log_it
+	traphandle default /usr/bin/perl /usr/bin/log_it
 
     In this case, it's almost certain that you'll also
   need to give the full path to the traphandle script (as shown)
@@ -1251,7 +1251,7 @@ OK, what should I put in snmp.conf?
   string to use.
 
     Some of these (such as the MIB file location), might be best put in
-  a shared snmp.conf file (typically /usr/local/share/snmp/snmp.conf or
+  a shared snmp.conf file (typically /usr/share/snmp/snmp.conf or
   /etc/snmp/snmp.conf) to apply to all users of the system.  Others
   (particularly the SNMPv3 security settings), are more likely to refer
   to a particular user, and should go in a personal snmp.conf file
@@ -1476,7 +1476,7 @@ Why aren't my mib files being read in?
 
 
     Alternatively, the tools may be looking in the wrong place.
-  The default location for the mib files is /usr/local/share/snmp/mibs.
+  The default location for the mib files is /usr/share/snmp/mibs.
   Again, this is set when the suite is first configured and compiled.
   This can be changed using the environmental variable 'MIBDIRS'
   or the snmp.conf directive 'mibdirs'.
@@ -2430,7 +2430,7 @@ How do I configure SNMPv3 users?
 
   There are three ways to configure SNMPv3 users:
 
-  1) Stop the agent, and create a file /var/net-snmp/snmpd.conf,
+  1) Stop the agent, and create a file /var/lib/snmp/snmpd.conf,
      containing the line
 
 	createUser {myUser} MD5 {myPassword} DES
@@ -2443,7 +2443,7 @@ How do I configure SNMPv3 users?
         net-snmp-config --create-snmpv3-user
 
      and follow the instructions.  This will create an entry
-     in the /var/net-snmp/snmpd.conf file similar to the above.
+     in the /var/lib/snmp/snmpd.conf file similar to the above.
      Then re-start the snmpd agent.
 
   3) Make sure the agent is running, and will respond to a suitable
@@ -2476,16 +2476,16 @@ The 'createUser' line disappears when I 
 
 
 
-What's the difference between /var/net-snmp and /usr/local/share/snmp?
----------------------------------------------------------------------
+What's the difference between /var/lib/snmp and /usr/share/snmp?
+----------------------------------------------------------------
 
-    The /var/net-snmp location is primarily used for information set
+    The /var/lib/snmp location is primarily used for information set
   during the running of the agent, which needs to be persistent between
   one run of the agent and the next.   Apart from "createUser" (see
   the previous entry), you shouldn't need to touch this file.
 
   All other user-provided configuration should go in the traditional
-  location (typically /usr/local/share/snmp/snmpd.conf or /etc/snmp).
+  location (typically /usr/share/snmp or /etc/snmp).
 
 
 
@@ -2498,7 +2498,7 @@ My new agent is ignoring the old snmpd.c
   distribution), with the current release installed from the source.
 
     The default location for this file with the basic distribution is
-  /usr/local/share/snmp/snmpd.conf (or PREFIX/share/snmp/snmpd.conf).
+  /usr/share/snmp/snmpd.conf (or PREFIX/share/snmp/snmpd.conf).
   Ready-installed versions often look for the file as /etc/snmpd.conf,
   or /etc/snmp/snmpd.conf.  Try moving the old config file to the new
   location, and restart the agent.
