Index: libshout-2.2.2/src/vorbis.c
===================================================================
--- libshout-2.2.2.orig/src/vorbis.c	2008-02-12 12:30:14.000000000 +0100
+++ libshout-2.2.2/src/vorbis.c	2008-02-12 12:30:14.000000000 +0100
@@ -27,7 +27,7 @@
 #include <inttypes.h>
 #endif
 
-#include <vorbis/codec.h>
+#include <tremor/ivorbiscodec.h>
 
 #include "shout_private.h"
 #include "shout_ogg.h"
