Index: classpath-0.98/configure.ac
===================================================================
--- classpath-0.98.orig/configure.ac	2009-02-06 00:26:08.000000000 +0100
+++ classpath-0.98/configure.ac	2012-08-22 10:18:53.325876508 +0200
@@ -762,8 +762,7 @@
   dnl change the name of the corresponding ac_ variable on lines 860...
   if test "x${COMPILE_GMP}" = xyes; then
     AC_CHECK_LIB(gmp, __gmpz_mul_si,
-      [GMP_CFLAGS=-I/usr/include
-       GMP_LIBS=-lgmp ],
+      [GMP_LIBS=-lgmp ],
       [GMP_CFLAGS=
        GMP_LIBS= ])
     AC_SUBST(GMP_CFLAGS)
