--- a/configure
+++ b/configure
@@ -24219,7 +24219,7 @@ if test "$cross_compiling" = yes; then
 See \`config.log' for more details." >&5
 $as_echo "$as_me: error: cannot run test program while cross compiling
 See \`config.log' for more details." >&2;}
-   { (exit 1); exit 1; }; }
+   }
 else
   cat >conftest.$ac_ext <<_ACEOF
 /* confdefs.h.  */
--- a/configure.in
+++ b/configure.in
@@ -196,24 +196,6 @@ dnl This generates warnings from autocon
 AC_C_BIGENDIAN
 
 
-dnl
-dnl   So many people with broken linking setups asked about this
-dnl it turned out to be necessary to check for this.
-dnl
-dofus_save_libs="$LIBS"
-LIBS="$ORBIT_LIBS $LIBS"
-AC_TRY_RUN([ #include <stdio.h>
-             main ()
-             {
-               return 0;
-             } ],,
-           AC_MSG_ERROR([
-Linking is badly borked on your system. Please ensure your library path is correct
-Check config.log for details - check near the end of the log above 'failed program was'
-Alternatively ensure that your /etc/ld.so.conf (and/or LD_LIBRARY_PATH) includes the
-prefix you're compiling on: '${prefix}' ]))
-LIBS="$dofus_save_libs"
-
 orig_CPPFLAGS=$CPPFLAGS
 CPPFLAGS="$CPPFLAGS $ORBIT_CFLAGS"
 AC_CHECK_ALIGNOF(CORBA_octet)
