diff -urN firmwarehotplug-0.1/ezusb/midi/original/Makefile firmwarehotplug-0.1.new/ezusb/midi/original/Makefile
--- firmwarehotplug-0.1/ezusb/midi/original/Makefile	2003-05-19 02:10:28.000000000 +0200
+++ firmwarehotplug-0.1.new/ezusb/midi/original/Makefile	2007-01-01 23:07:42.000000000 +0100
@@ -1,7 +1,7 @@
 VERSION = $(shell date +%Y_%m_%d)
 RELEASE_NAME =	ezusbmidi-$(VERSION)
 
-all: ezusbmidi1x1.ihx ezusbmidi2x2.ihx ezusbmidi.html ezusbmidi.ps
+all: ezusbmidi1x1.ihx ezusbmidi2x2.ihx
 
 ezusbmidi1x1.ihx: ezusbmidi.c ezusb_reg.h bufsync.h
 	rm -rf sdcc; mkdir sdcc; cp $^ sdcc
