--- a/ext/standard/basic_functions.c
+++ b/ext/standard/basic_functions.c
@@ -436,7 +436,7 @@ function_entry basic_functions[] = {
 	PHP_FE(gethostbyname,													NULL)
 	PHP_FE(gethostbynamel,													NULL)
 
-#if HAVE_RES_SEARCH && !(defined(__BEOS__) || defined(PHP_WIN32) || defined(NETWARE))
+#if HAVE_RES_SEARCH && HAVE_DN_SKIPNAME && !(defined(__BEOS__)||defined(PHP_WIN32) || defined(NETWARE))
 	PHP_FE(checkdnsrr,														NULL)
 	PHP_FE(getmxrr,second_and_third_args_force_ref)
 #endif
--- a/ext/standard/dns.c
+++ b/ext/standard/dns.c
@@ -207,7 +207,7 @@ static char *php_gethostbyname(char *nam
 }
 /* }}} */
 
-#if HAVE_RES_SEARCH && !(defined(__BEOS__)||defined(PHP_WIN32) || defined(NETWARE))
+#if HAVE_RES_SEARCH && HAVE_DN_SKIPNAME && !(defined(__BEOS__)||defined(PHP_WIN32) || defined(NETWARE))
 
 /* {{{ proto int checkdnsrr(string host [, string type])
    Check DNS records corresponding to a given Internet host name or IP address */
