--- a/Zend/zend_strtod.c
+++ b/Zend/zend_strtod.c
@@ -117,6 +117,7 @@ typedef unsigned long int uint32_t;
 # endif
 #endif
 
+#if !defined(IEEE_BIG_ENDIAN) && !defined(IEEE_LITTLE_ENDIAN)
 #ifdef WORDS_BIGENDIAN
 #define IEEE_BIG_ENDIAN
 #else
@@ -141,6 +142,7 @@ typedef unsigned long int uint32_t;
 #define uint32_t unsigned __int32
 #define IEEE_LITTLE_ENDIAN
 #endif
+#endif /* !defined(IEEE_BIG_ENDIAN) && !defined(IEEE_LITTLE_ENDIAN) */
 
 #define Long	int32_t
 #define ULong	uint32_t
