--- a/kernel/wctdm.c
+++ b/kernel/wctdm.c
@@ -30,6 +30,8 @@
 #include <linux/errno.h>
 #include <linux/pci.h>
 #include <linux/interrupt.h>
+#include <linux/slab.h>
+#include <linux/sched.h>
 #include <asm/io.h>
 #include "proslic.h"
 #include "wctdm.h"
--- a/kernel/wcusb.c
+++ b/kernel/wcusb.c
@@ -37,6 +37,7 @@
 #include <linux/init.h>
 #include <linux/usb.h>
 #include <linux/errno.h>
+#include <linux/slab.h>
 
 #include <linux/version.h>
 #if LINUX_VERSION_CODE > KERNEL_VERSION(2,4,19)
--- a/kernel/ztdummy.c
+++ b/kernel/ztdummy.c
@@ -38,6 +38,7 @@
  */
 
 #include <linux/version.h>
+#include <linux/slab.h>
 
 #ifndef VERSION_CODE
 #  define VERSION_CODE(vers,rel,seq) ( ((vers)<<16) | ((rel)<<8) | (seq) )
