--- a/kernel/ztdummy.c
+++ b/kernel/ztdummy.c
@@ -79,6 +79,7 @@
 #include <asm/io.h>
 #endif
 #ifdef LINUX26
+#include <linux/spinlock.h>
 #ifdef USE_HIGHRESTIMER
 #include <linux/hrtimer.h>
 #endif
@@ -328,7 +329,7 @@ int init_module(void)
 #ifdef LINUX26
 	ztd->counter = 0;
 #ifdef USE_RTC
-	ztd->rtclock = SPIN_LOCK_UNLOCKED;
+	spin_lock_init(&ztd->rtclock);
 	ztd->rtc_task.func = ztdummy_rtc_interrupt;
 	ztd->rtc_task.private_data = ztd;
 	err = rtc_register(&ztd->rtc_task);
