diff -urN libdbi-0.8.3/Makefile.am libdbi-0.8.3.new/Makefile.am
--- libdbi-0.8.3/Makefile.am	2006-05-18 22:27:50.000000000 +0200
+++ libdbi-0.8.3.new/Makefile.am	2009-01-09 00:06:34.000000000 +0100
@@ -2,7 +2,7 @@
 
 AUTOMAKE_OPTIONS = foreign dist-zip
 
-SUBDIRS = src include @docs_subdirs@
+SUBDIRS = src include
 
 EXTRA_DIST = AUTHORS COPYING ChangeLog INSTALL README README.win32 README.osx TODO NEWS libdbi.spec.in config.h.in Makefile.w32 autogen.sh drivers/example/AUTHORS drivers/example/Makefile.am drivers/example/README drivers/example/TODO drivers/example/.cvsignore drivers/example/dbd_example.c drivers/example/dbd_example.h drivers/example/dbd_example.sgml
 
diff -urN libdbi-0.8.3/Makefile.in libdbi-0.8.3.new/Makefile.in
--- libdbi-0.8.3/Makefile.in	2008-02-06 20:40:10.000000000 +0100
+++ libdbi-0.8.3.new/Makefile.in	2009-01-09 00:06:28.000000000 +0100
@@ -180,7 +180,7 @@
 sysconfdir = @sysconfdir@
 target_alias = @target_alias@
 AUTOMAKE_OPTIONS = foreign dist-zip
-SUBDIRS = src include @docs_subdirs@
+SUBDIRS = src include
 EXTRA_DIST = AUTHORS COPYING ChangeLog INSTALL README README.win32 README.osx TODO NEWS libdbi.spec.in config.h.in Makefile.w32 autogen.sh drivers/example/AUTHORS drivers/example/Makefile.am drivers/example/README drivers/example/TODO drivers/example/.cvsignore drivers/example/dbd_example.c drivers/example/dbd_example.h drivers/example/dbd_example.sgml
 all: config.h
 	$(MAKE) $(AM_MAKEFLAGS) all-recursive
