--- a/Makefile
+++ b/Makefile
@@ -501,6 +501,16 @@ install-include: zaptel.h torisa.h
 	install -d $(DESTDIR)$(INCLUDE_DIR)/linux
 	install -m 644 zaptel.h torisa.h $(DESTDIR)$(INCLUDE_DIR)/linux/
 
+install-tonezone: $(LIBTONEZONE_SO)
+	install -D -m 755 $(LIBTONEZONE_SO) $(DESTDIR)/usr/lib/$(LIBTONEZONE_SO).$(LIBTONEZONE_SO_MAJOR_VER).$(LIBTONEZONE_SO_MINOR_VER)
+	rm -f $(DESTDIR)/usr/lib/$(LIBTONEZONE_SO)
+	ln -sf $(LIBTONEZONE_SO).$(LIBTONEZONE_SO_MAJOR_VER).$(LIBTONEZONE_SO_MINOR_VER) \
+		$(DESTDIR)/usr/lib/$(LIBTONEZONE_SO).$(LIBTONEZONE_SO_MAJOR_VER)
+	ln -sf $(LIBTONEZONE_SO).$(LIBTONEZONE_SO_MAJOR_VER).$(LIBTONEZONE_SO_MINOR_VER) \
+		$(DESTDIR)/usr/lib/$(LIBTONEZONE_SO)
+	install -D -m 644 zaptel.h $(DESTDIR)/usr/include/linux/zaptel.h
+	install -D -m 644 tonezone.h $(DESTDIR)/usr/include/tonezone.h
+
 install-udev: devices
 
 # FIRMWARE_DIR also implies HOTPLUG_FIRMWARE, unless set manually.
