--- xl2tpd-1.3.1/control.c.orig	2012-04-25 12:59:28.718825985 +0100
+++ xl2tpd-1.3.1/control.c	2012-04-25 12:59:37.063916785 +0100
@@ -905,8 +905,6 @@ int control_finish (struct tunnel *t, st
                 po = add_opt (po, c->lac->pppoptfile);
             }
         };
-	po = add_opt (po, "ipparam");
-        po = add_opt (po, IPADDY (t->peer.sin_addr));
         start_pppd (c, po);
         opt_destroy (po);
         if (c->lac)
