--- a/prosody.cfg.lua.dist
+++ b/prosody.cfg.lua.dist
@@ -47,7 +47,7 @@ modules_enabled = {
 		"register"; -- Allow users to register on this server using a client and change passwords
 	
 	-- Other specific functionality
-		--"posix"; -- POSIX functionality, sends server to background, enables syslog, etc.
+		"posix"; -- POSIX functionality, sends server to background, enables syslog, etc.
 		--"console"; -- Opens admin telnet interface on localhost port 5582
 		--"bosh"; -- Enable BOSH clients, aka "Jabber over HTTP"
 		--"httpserver"; -- Serve static files from a directory over HTTP
@@ -65,6 +65,9 @@ modules_disabled = {
 -- Disable account creation by default, for security
 -- For more information see http://prosody.im/doc/creating_accounts
 allow_registration = false;
+
+-- File to write pid in
+pidfile = "/var/run/prosody/prosody.pid";
 	
 -- These are the SSL/TLS-related settings. If you don't want
 -- to use SSL/TLS, you may comment or remove this
@@ -73,6 +76,15 @@ ssl = {
 	certificate = "certs/localhost.cert";
 }
 
+-- Errors to syslog
+-- All to /var/log/prosody/
+log = {
+	{ levels = { "error" }; to = "syslog";  };
+	{ levels = { "error" }; to = "file"; filename = "/var/log/prosody/prosody.err";  };
+	-- Change "info" to "debug" for more verbose logging
+	{ levels = { min = "info" }; to = "file"; filename = "/var/log/prosody/prosody.log";  };
+}
+
 ----------- Virtual hosts -----------
 -- You need to add a VirtualHost entry for each domain you wish Prosody to serve.
 -- Settings under each VirtualHost entry apply *only* to that host.
