From 525bfb210c381fade8489ec54efbed8b2897fac3 Mon Sep 17 00:00:00 2001
From: Seif Mazareeb <seif@marvell.com>
Date: Mon, 5 Mar 2012 18:17:32 +0200
Subject: [PATCH 107/609] DSMP adding and ebaling NEED_MACH_MEMORY_H

Signed-off-by: Seif Mazareeb <seif@marvell.com>
---
 arch/arm/Kconfig |    1 +
 1 file changed, 1 insertion(+)

--- a/arch/arm/Kconfig
+++ b/arch/arm/Kconfig
@@ -578,6 +578,7 @@ config ARCH_ARMADA_XP
 	select COMMON_CLKDEV
 	select CLKSRC_MMIO
 	select CLKDEV_LOOKUP
+	select NEED_MACH_MEMORY_H
 
 config ARCH_MV78XX0
 	bool "Marvell MV78xx0"
