From 9da8c3b74ab4b139da70879329d416992b6b10c3 Mon Sep 17 00:00:00 2001
From: Seif Mazareeb <seif@marvell.com>
Date: Sun, 12 Feb 2012 10:45:29 +0200
Subject: [PATCH 123/609] DSMP adding AFLAGS to sheeva v7 / v6

Signed-off-by: Seif Mazareeb <seif@marvell.com>
---
 arch/arm/mm/Makefile |    3 +++
 1 file changed, 3 insertions(+)

--- a/arch/arm/mm/Makefile
+++ b/arch/arm/mm/Makefile
@@ -99,6 +99,9 @@ obj-$(CONFIG_CPU_SHEEVA_PJ4B_V7)	+= proc
 AFLAGS_proc-v6.o	:=-Wa,-march=armv6
 AFLAGS_proc-v7.o	:=-Wa,-march=armv7-a
 
+AFLAGS_proc-sheeva_pj4bv7.o        :=-Wa,-march=armv6
+AFLAGS_proc-sheeva_pj4bv7.o        :=-Wa,-march=armv7-a
+
 obj-$(CONFIG_CACHE_FEROCEON_L2)	+= cache-feroceon-l2.o
 obj-$(CONFIG_CACHE_L2X0)	+= cache-l2x0.o
 obj-$(CONFIG_CACHE_XSC3L2)	+= cache-xsc3l2.o
