From 11b234325e2d29649fb15e0c779e9ac798cf3642 Mon Sep 17 00:00:00 2001
From: Seif Mazareeb <seif@marvell.com>
Date: Tue, 12 Jun 2012 16:37:47 +0300
Subject: [PATCH 143/609] fixed for R.H defconfig

Signed-off-by: Seif Mazareeb <seif@marvell.com>
---
 include/linux/miscdevice.h |    2 ++
 1 file changed, 2 insertions(+)

--- a/include/linux/miscdevice.h
+++ b/include/linux/miscdevice.h
@@ -39,11 +39,13 @@
 #define MWAVE_MINOR		219	/* ACP/Mwave Modem */
 #define MPT_MINOR		220
 #define MPT2SAS_MINOR		221
+#define UINPUT_MINOR           	223
 #define HPET_MINOR		228
 #define FUSE_MINOR		229
 #define KVM_MINOR		232
 #define BTRFS_MINOR		234
 #define AUTOFS_MINOR		235
+#define MAPPER_CTRL_MINOR       236
 #define LOOP_CTRL_MINOR		237
 #define MISC_DYNAMIC_MINOR	255
 
