From 35cfc148dbb3629a74372b52174e28775c4d2e2c Mon Sep 17 00:00:00 2001
From: Dmitri Epshtein <dima@marvell.com>
Date: Mon, 9 Jul 2012 10:09:41 -0400
Subject: [PATCH 165/609] rename NFP hook

Signed-off-by: Seif Mazareeb <seif@marvell.com>
---
 net/core/dev.c |    2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

--- a/net/core/dev.c
+++ b/net/core/dev.c
@@ -6675,7 +6675,7 @@ static int __init net_dev_init(void)
 	dev_mcast_init();
 	rc = 0;
 #if defined(CONFIG_MV_ETH_NFP) || defined(CONFIG_MV_ETH_NFP_MODULE)
-	init_nfp_hook_ops();
+	nfp_init_nfp_hook_ops();
 #endif
 out:
 	return rc;
