From 0ff267f111d1aa6bc2e69af67ad54926839cd835 Mon Sep 17 00:00:00 2001
From: Rami Rosen <rosenr@marvell.com>
Date: Wed, 16 May 2012 15:20:34 +0300
Subject: [PATCH 188/609] fix build NFP modules w/wo Netfilter

Signed-off-by: Seif Mazareeb <seif@marvell.com>
---
 include/linux/mv_nfp.h |   10 ++++++++++
 1 file changed, 10 insertions(+)

--- a/include/linux/mv_nfp.h
+++ b/include/linux/mv_nfp.h
@@ -66,8 +66,13 @@ SOFTWARE, EVEN IF ADVISED OF THE POSSIBI
 #ifndef LINUX_MV_NFP_H
 #define LINUX_MV_NFP_H
 
+#ifdef CONFIG_NETFILTER
 #include <net/netfilter/nf_conntrack.h>
 #include <net/netfilter/nf_nat.h>
+#endif
+#include <linux/types.h>
+#include <linux/netdevice.h>
+
 
 #if defined(CONFIG_MV_ETH_NFP) || defined(CONFIG_MV_ETH_NFP_MODULE)
 
@@ -119,9 +124,14 @@ struct nfp_hook_ops {
 
 	int (*nfp_hook_ct_del)(int family, u8 *src_l3, u8 *dst_l3, u16 sport, u16 dport, u8 proto);
 	int (*nfp_hook_ct_age)(int family, u8 *src_l3, u8 *dst_l3, u16 sport, u16 dport, u8 proto);
+#ifdef CONFIG_NETFILTER
 	int (*nfp_hook_ct_nat_add)(const struct nf_conntrack_tuple *tuple,
 				const struct nf_conntrack_tuple *target, enum nf_nat_manip_type maniptype);
 	int  (*nfp_hook_ct_fwd_add)(const struct nf_conntrack_tuple *tuple, int mode);
+#else
+	int (*nfp_hook_ct_nat_add)(void);
+	int (*nfp_hook_ct_fwd_add)(void);
+#endif
 	int  (*nfp_ct_is_learning_enabled)(void);
 };
 
