From bb7990c84f28651d81e4399a9d3ad8da637949d0 Mon Sep 17 00:00:00 2001
From: Lennox Wu <lennoxwu@marvell.com>
Date: Wed, 1 Feb 2012 18:47:18 +0800
Subject: [PATCH 229/609] Fix the Makefile for LPAE

Signed-off-by: Seif Mazareeb <seif@marvell.com>
---
 arch/arm/mm/Makefile |    2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

--- a/arch/arm/mm/Makefile
+++ b/arch/arm/mm/Makefile
@@ -96,7 +96,7 @@ obj-$(CONFIG_CPU_FEROCEON)	+= proc-feroc
 #obj-$(CONFIG_CPU_V7)		+= proc-v7.o
 obj-$(CONFIG_CPU_SHEEVA_PJ4B_V6)	+= proc-sheeva_pj4bv6.o
 ifeq ($(CONFIG_ARM_LPAE),y)
-obj-$(CONFIG_CPU_V7)		+= proc-v7lpae.o
+obj-$(CONFIG_CPU_SHEEVA_PJ4B_V7)	+= proc-sheeva_pj4bv7lpae.o
 else
 obj-$(CONFIG_CPU_SHEEVA_PJ4B_V7)	+= proc-sheeva_pj4bv7.o
 endif
