From c64d27fbe1ba9fe5a1c09200b8d845dc1298d0a0 Mon Sep 17 00:00:00 2001
From: Yehuda Yitschak <yehuday@marvell.com>
Date: Wed, 31 Oct 2012 13:42:26 +0200
Subject: [PATCH 331/609] Fixed hwmon Kconfig to support KW40 and AXP

Signed-off-by: Yehuda Yitschak <yehuday@marvell.com>
Signed-off-by: Seif Mazareeb <seif@marvell.com>
---
 drivers/hwmon/Kconfig |    4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

--- a/drivers/hwmon/Kconfig
+++ b/drivers/hwmon/Kconfig
@@ -220,9 +220,9 @@ config SENSORS_ADT7475
 
 	  This driver can also be build as a module.  If so, the module
 	  will be called adt7475.
-config SENSORS_ARMADA_XP
+config SENSORS_ARMADA
 	tristate "Armada XP SoC Thermal Sensor"
-	depends on ARCH_ARMADA_XP
+	depends on PLAT_ARMADA
 	default y
 	help
 	  This driver provides support for the Marvell's Armada XP SoC thermal
