From f1599139f8695ed64f65956997548adbae7236dc Mon Sep 17 00:00:00 2001
From: Po-Yu Chuang <pychuang@marvell.com>
Date: Wed, 17 Oct 2012 10:31:33 +0800
Subject: [PATCH 336/609] -march=armv6 is for AFLAGS_proc-sheeva_pj4bv6.o

Signed-off-by: Shadi Ammouri <shadi@marvell.com>
Signed-off-by: Seif Mazareeb <seif@marvell.com>
---
 arch/arm/mm/Makefile |    2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

--- a/arch/arm/mm/Makefile
+++ b/arch/arm/mm/Makefile
@@ -106,7 +106,7 @@ endif
 AFLAGS_proc-v6.o	:=-Wa,-march=armv6
 AFLAGS_proc-v7.o	:=-Wa,-march=armv7-a
 
-AFLAGS_proc-sheeva_pj4bv7.o        :=-Wa,-march=armv6
+AFLAGS_proc-sheeva_pj4bv6.o        :=-Wa,-march=armv6
 AFLAGS_proc-sheeva_pj4bv7.o        :=-Wa,-march=armv7-a
 
 obj-$(CONFIG_CACHE_FEROCEON_L2)	+= cache-feroceon-l2.o
