From 246dbcdd1104fadcccc085c148232848f01a53ff Mon Sep 17 00:00:00 2001
From: Po-Yu Chuang <pychuang@marvell.com>
Date: Wed, 17 Oct 2012 10:41:42 +0800
Subject: [PATCH 337/609] add AFLAGS_proc-sheeva_pj4bv7lpae.o

Signed-off-by: Shadi Ammouri <shadi@marvell.com>
Signed-off-by: Seif Mazareeb <seif@marvell.com>
---
 arch/arm/mm/Makefile |    1 +
 1 file changed, 1 insertion(+)

--- a/arch/arm/mm/Makefile
+++ b/arch/arm/mm/Makefile
@@ -108,6 +108,7 @@ AFLAGS_proc-v7.o	:=-Wa,-march=armv7-a
 
 AFLAGS_proc-sheeva_pj4bv6.o        :=-Wa,-march=armv6
 AFLAGS_proc-sheeva_pj4bv7.o        :=-Wa,-march=armv7-a
+AFLAGS_proc-sheeva_pj4bv7lpae.o    :=-Wa,-march=armv7-a
 
 obj-$(CONFIG_CACHE_FEROCEON_L2)	+= cache-feroceon-l2.o
 obj-$(CONFIG_CACHE_L2X0)	+= cache-l2x0.o
