From cd98f46b1638de56a33367582e88b9b8d86d7476 Mon Sep 17 00:00:00 2001
From: Nadav Haklai <nadavh@marvell.com>
Date: Mon, 7 Jan 2013 14:30:50 +0200
Subject: [PATCH 392/609] Fix: hwmon compilation fix for AXP

Signed-off-by: Seif Mazareeb <seif@marvell.com>
---
 arch/arm/mach-armadaxp/Makefile |    2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

--- a/arch/arm/mach-armadaxp/Makefile
+++ b/arch/arm/mach-armadaxp/Makefile
@@ -135,7 +135,7 @@ obj-$(CONFIG_MV_IPC_NET)		+= $(PLAT_DRIV
 obj-y					+= $(PLAT_DRIVERS)/mv_gpio/
 obj-$(CONFIG_MV_DBG_TRACE)              += $(PLAT_DRIVERS)/mv_trace/
 obj-$(CONFIG_MV_INCLUDE_SWITCH)         += $(LSP_SWITCH_DIR)/
-obj-$(CONFIG_SENSORS_ARMADA_XP)		+= hwmon.o
+obj-$(CONFIG_SENSORS_ARMADA)		+= hwmon.o
 # The rest of the drivers are compiled through the driver dir directly.
 
 
