--- a/arch/arm/plat-armada/mv_hal/pex/mvPex.c
+++ b/arch/arm/plat-armada/mv_hal/pex/mvPex.c
@@ -96,6 +96,11 @@ MV_STATUS mvPexInit(MV_U32 pexIf, MV_PEX
 		return MV_ERROR;
 	}
 
+	if (pexMode.pexLinkUp == MV_FALSE) {
+		/*  interface detected no Link. */
+		return MV_NO_SUCH;
+	}
+
 	/* Check that required PEX type is the one set in reset time */
 	if (pexType != pexMode.pexType) {
 		/* No Link. Shut down the Phy */
@@ -117,44 +122,11 @@ MV_STATUS mvPexInit(MV_U32 pexIf, MV_PEX
 		status = MV_REG_READ(PEX_CFG_DIRECT_ACCESS(pexIf, PEX_STATUS_AND_COMMAND));
 		status |= PXSAC_INT_DIS;
 		MV_REG_WRITE(PEX_CFG_DIRECT_ACCESS(pexIf, PEX_STATUS_AND_COMMAND), status);
-		if (ctrlFamily==MV_67XX) {
-			/* PEX capability */
-			regVal = MV_REG_READ(PEX_CFG_DIRECT_ACCESS(pexIf, PEX_LINK_CAPABILITY_REG)) & 0xF;
-			if (regVal == 0x2) {
-				/* KW40 set to Gen 2.0 - conf_auto_speed_change */
-				/* when bit is set, link will issue link speed change to the max link speed possible */
-				regVal = MV_REG_READ(PEX_CTRL_REG(pexIf)) | (1<<10);
-				MV_REG_WRITE(PEX_CTRL_REG(pexIf), regVal);
-			}
-		}
-	} else { /* if (MV_PEX_ROOT_COMPLEX != pexType) */
-		if (ctrlFamily==MV_67XX) {
-			/* TODO: 14/12/10 - requested by CV to support EP Compliance */
-			MV_REG_WRITE(PEX_DBG_CTRL_REG(pexIf), 0x0F62F0C0);
-
-			MV_REG_WRITE(PEX_PHY_ACCESS_REG(pexIf), (0x80C2 << 16));
-			regVal = MV_REG_READ(PEX_PHY_ACCESS_REG(pexIf)) & 0xFFFF;
-			MV_REG_WRITE(PEX_PHY_ACCESS_REG(pexIf), (0xC2 << 16) | regVal | (1 << 2));
-
-			MV_REG_WRITE(PEX_PHY_ACCESS_REG(pexIf), (0x8080 << 16));
-			regVal = MV_REG_READ(PEX_PHY_ACCESS_REG(pexIf)) & 0xFFFF;
-			MV_REG_WRITE(PEX_PHY_ACCESS_REG(pexIf), (0x80 << 16) | regVal | (0x2 << 1));
-		} else {  /* if Armada-XP : */
+	} else {
 			regVal = MV_REG_READ(PEX_DBG_CTRL_REG(pexIf));
 			regVal &= ~(BIT16 | BIT19);
 			MV_REG_WRITE(PEX_DBG_CTRL_REG(pexIf), regVal);
-		}
-	}
 
-	if (ctrlFamily==MV_67XX) {
-		 mvCpuIfEnablePex(pexIf);
-	}
-	/* now wait 1ms to be sure the link is valid */
-	mvOsDelay(1);
-	/* Check if we have link */
-	if (MV_REG_READ(PEX_STATUS_REG(pexIf)) & PXSR_DL_DOWN) {
-		/*mvOsPrintf("PEX%d interface detected no Link.\n", pexIf);*/
-		return MV_NO_SUCH;
 	}
 #ifdef PCIE_VIRTUAL_BRIDGE_SUPPORT
 	mvPexVrtBrgInit(pexIf);
@@ -184,7 +156,7 @@ MV_U32 mvPexModeGet(MV_U32 pexIf, MV_PEX
 	if (pexIf >= MV_PEX_MAX_IF)
 		return MV_BAD_PARAM;
 
-
+	
 	pexData = MV_REG_READ(PEX_CTRL_REG(pexIf));
 
 	switch (pexData & PXCR_DEV_TYPE_CTRL_MASK) {
