From 812375b16c7bc2bab427317d0b29cb336aa80bf4 Mon Sep 17 00:00:00 2001
From: Tawfik Bayouk <tawfik@marvell.com>
Date: Wed, 16 Jan 2013 17:31:44 +0200
Subject: [PATCH 504/609] Bump Quarter Baseline 2013_Q1

Signed-off-by: Seif Mazareeb <seif@marvell.com>
---
 version.txt |    1 +
 1 file changed, 1 insertion(+)
 create mode 100644 version.txt

--- /dev/null
+++ b/version.txt
@@ -0,0 +1 @@
+2013_Q1.0
