--- a/arch/arm/mach-armadaxp/Makefile
+++ b/arch/arm/mach-armadaxp/Makefile
@@ -104,7 +104,6 @@ armadaxp-$(CONFIG_MV_INCLUDE_AUDIO) 	+=
 armadaxp-$(CONFIG_MV_INCLUDE_TS) 	+= $(HAL_TS_DIR)/mvTsu.o $(HAL_IF_DIR)/mvSysTs.o		\
 					   $(HAL_TS_DIR)/mvTsuAddrDec.o
 armadaxp-$(CONFIG_MV_CPU_PERF_CNTRS)    += $(HAL_CPU_DIR)/mvCpuCntrs.o $(HAL_CPU_DIR)/pj4/mvPJ4Cntrs.o
-armadaxp-$(CONFIG_PCIE_VIRTUAL_BRIDGE_SUPPORT) += $(HAL_PEX_DIR)/mvVrtBrgPex.o
 armadaxp-$(CONFIG_MV_CPU_L2_PERF_CNTRS) += $(HAL_CPU_DIR)/mvCpuL2Cntrs.o
 
 obj-$(CONFIG_MV_INCLUDE_SWITCH) 	+= $(QD_OBJS)
--- a/arch/arm/mach-armadaxp/config/mvRules.mk
+++ b/arch/arm/mach-armadaxp/config/mvRules.mk
@@ -172,10 +172,6 @@ ifeq ($(CONFIG_MV_SATA_ENABLE_1MB_IOS),y
 EXTRA_CFLAGS    += -DMV_SUPPORT_1MBYTE_IOS
 endif
 
-ifeq ($(CONFIG_PCIE_VIRTUAL_BRIDGE_SUPPORT),y)
-EXTRA_CFLAGS    +=-DPCIE_VIRTUAL_BRIDGE_SUPPORT
-endif
-
 ifeq ($(CONFIG_MV_CESA_CHAIN_MODE_SUPPORT),y)
 EXTRA_CFLAGS    += -DMV_CESA_CHAIN_MODE_SUPPORT
 endif
--- a/arch/arm/plat-armada/Kconfig
+++ b/arch/arm/plat-armada/Kconfig
@@ -521,12 +521,6 @@ config  SCSI_MVSATA
 
 source arch/arm/plat-armada/mv_drivers_lsp/mv_sata/Kconfig
 
-config PCIE_VIRTUAL_BRIDGE_SUPPORT
-        bool "Enable virtual bridge on PCIe"
-        depends on PCI
-        ---help---
-        Enable virtual bridge module to support PCIe switches
-
 config MV_UART_POLLING_MODE
         bool "Enable Polling mode for UART driver"
         depends on ARCH_FEROCEON_KW2
