--- a/arch/arm/plat-armada/Kconfig
+++ b/arch/arm/plat-armada/Kconfig
@@ -157,7 +157,7 @@ config MV_INCLUDE_SPI
 
 config MV_INCLUDE_NOR
 	bool "NOR Support"
-	depends on MV88F6500 || MV88F6281 || (MV78XX0 && !MV78XX0_Z0) || ARMADA_XP
+	depends on MV88F6500 || MV88F6281 || (MV78XX0 && !MV78XX0_Z0) || ARMADA_XP || (ARMADA_370 && !MV_INCLUDE_NFC)
 	default y
         ---help---
         Please don't change this configs unless you know what you are doing.
