--- a/arch/arm/boot/compressed/head.S
+++ b/arch/arm/boot/compressed/head.S
@@ -12,6 +12,9 @@
 #ifdef CONFIG_ARCH_ARMADA_XP
 #include <mach/armadaxp.h>
 #endif
+#ifdef CONFIG_ARCH_ARMADA370
+#include <mach/armada370.h>
+#endif
 
 /*
  * Debugging stuff
@@ -139,12 +142,23 @@ start:
 		.word	_edata			@ zImage end address
  THUMB(		.thumb			)
 1:
+#ifdef CONFIG_PLAT_ARMADA
+/* Update Internal Regs offset in case UBoot is configured
+** to use a different base address.
+*/
+		mrc p15, 0, r0, c5, c0, 0	@ Get the internal registers indication
+		tst r0, #(1 << 11)		@ Check 0xF1000000 indication - bit 11
+		beq 1f				@ If bit 11 is not set (equal):
+						@ AXP - Need to check original Internal regs offset CP15 register
+						@ A370 - U-Boot was using 0xD0000000
+		b 2f				@ Else, bit 11 is set - U-Boot was using 0xF1000000 - Nothing to do
+1:
 #ifdef CONFIG_ARCH_ARMADA_XP
-		/* Update Internal Regs offset in case UBoot is configured
-		** to use a different base address.
-		*/
 		mrc p15, 4, r0, c15, c0, 0	@ Get the internal registers base address
 		lsl r0, r0, #13			@ the address is R-shifted, need to recover it
+#else	/* CONFIG_ARCH_ARMADA_370 */
+		ldr r0, =0xD0000000
+#endif
 		ldr r5, =0x20080
 		add r0, r0, r5
 		ldr r6, =INTER_REGS_PHYS_BASE
@@ -152,6 +166,7 @@ start:
 		rev r6, r6
 #endif
 		str r6, [r0]
+2:
 #endif
 
 		mov	r7, r1			@ save architecture ID
