--- a/include/net/netfilter/nf_conntrack.h
+++ b/include/net/netfilter/nf_conntrack.h
@@ -128,21 +128,6 @@ struct nf_conn {
 	u_int32_t secmark;
 #endif
 
-#if defined(CONFIG_NETFILTER_XT_MATCH_LAYER7) || \
-    defined(CONFIG_NETFILTER_XT_MATCH_LAYER7_MODULE)
-	struct {
-		/*
-		 * e.g. "http". NULL before decision. "unknown" after decision
-		 * if no match.
-		 */
-		char *app_proto;
-		/*
-		 * application layer data so far. NULL after match decision.
-		 */
-		char *app_data;
-		unsigned int app_data_len;
-	} layer7;
-#endif
 	/* Extensions */
 	struct nf_ct_ext *ext;
 #ifdef CONFIG_NET_NS
