--- a/arch/arm/mach-armadaxp/armada_xp_family/boardEnv/mvBoardEnvSpec.c
+++ b/arch/arm/mach-armadaxp/armada_xp_family/boardEnv/mvBoardEnvSpec.c
@@ -1408,8 +1408,8 @@ MV_BOARD_INFO rd78460wifiInfo = {
         .gppPolarityValHigh             = RD_78460_WIFI_GPP_POL_HIGH,
 
         /* External Switch Configuration */
-        .pSwitchInfo = NULL,
-        .switchInfoNum = 0,
+        .pSwitchInfo = rd78460gpBoardSwitchValue,
+        .switchInfoNum = 1,
 
         /* TDM configuration */
         /* We hold a different configuration array for each possible slic that
