--- a/arch/arm/mach-armadaxp/board_mamba.c
+++ b/arch/arm/mach-armadaxp/board_mamba.c
@@ -44,6 +44,7 @@ disclaimer.
  * */
 #define	MAMBA_GPIO_WPS_BUTTON	32
 #define MAMBA_GPIO_RST_BUTTON	33
+#define MAMBA_GPIO_POWER_LED	40
 
 static struct gpio_keys_button mv_ev_buttons[] = {
         {
@@ -186,6 +187,7 @@ static int __init mamba_ev_init(void)
 
 	orion_gpio_set_valid(MAMBA_GPIO_WPS_BUTTON, 1);
 	orion_gpio_set_valid(MAMBA_GPIO_RST_BUTTON, 1);
+	orion_gpio_set_valid(MAMBA_GPIO_POWER_LED, 1);
 
 	/* register device */
 	status = platform_device_register(&mamba_ev_device);
