diff -Naur linux-3.2.36_org/arch/arm/mm/Kconfig linux-3.2.36/arch/arm/mm/Kconfig
--- linux-3.2.36_org/arch/arm/mm/Kconfig	2014-07-13 13:14:50.830212086 +0700
+++ linux-3.2.36/arch/arm/mm/Kconfig	2014-07-13 13:17:05.162212371 +0700
@@ -838,7 +838,7 @@
 	  v6 mode where cache maintenance breadcasting done in sotfware.
 
 config OUTER_CACHE
-	bool
+	bool "Enable outer_cache" 
 
 config OUTER_CACHE_SYNC
 	bool
