--- a/configure
+++ b/configure
@@ -14776,7 +14776,7 @@ fi
 
 
 
-VPZONE_CFLAGS="-std=gnu99 -Wall -Werror -Wextra -fdiagnostics-show-option ${DEV_CFLAGS}"
+VPZONE_CFLAGS="-std=gnu99 -Wall -Wextra -fdiagnostics-show-option ${DEV_CFLAGS}"
 
 VPZONE_LDFLAGS="-ldl -levent -lcares"
 
--- a/configure.ac
+++ b/configure.ac
@@ -65,7 +65,7 @@ dnl # #######################
 dnl Compiler and Linker flags
 dnl # #######################
 
-VPZONE_CFLAGS="-std=gnu99 -Wall -Werror -Wextra -fdiagnostics-show-option ${DEV_CFLAGS}"
+VPZONE_CFLAGS="-std=gnu99 -Wall -Wextra -fdiagnostics-show-option ${DEV_CFLAGS}"
 AC_SUBST(VPZONE_CFLAGS)
 VPZONE_LDFLAGS="-ldl -levent -lcares"
 AC_SUBST(VPZONE_LDFLAGS)
