--- a/addons/app_mysql.c
+++ b/addons/app_mysql.c
@@ -25,7 +25,7 @@
 
 /*** MODULEINFO
 	<depend>mysqlclient</depend>
-	<defaultenabled>no</defaultenabled>
+	<defaultenabled>yes</defaultenabled>
 	<support_level>deprecated</support_level>
 	<replacement>func_odbc</replacement>
  ***/
--- a/addons/app_saycountpl.c
+++ b/addons/app_saycountpl.c
@@ -21,7 +21,7 @@
  */
 
 /*** MODULEINFO
-	<defaultenabled>no</defaultenabled>
+	<defaultenabled>yes</defaultenabled>
 	<support_level>deprecated</support_level>
 	<replacement>say.conf</replacement>
  ***/
--- a/addons/cdr_mysql.c
+++ b/addons/cdr_mysql.c
@@ -35,7 +35,7 @@
 
 /*** MODULEINFO
 	<depend>mysqlclient</depend>
-	<defaultenabled>no</defaultenabled>
+	<defaultenabled>yes</defaultenabled>
 	<support_level>deprecated</support_level>
 	<replacement>cdr_adaptive_odbc</replacement>
  ***/
--- a/addons/chan_mobile.c
+++ b/addons/chan_mobile.c
@@ -27,7 +27,7 @@
 
 /*** MODULEINFO
 	<depend>bluetooth</depend>
-	<defaultenabled>no</defaultenabled>
+	<defaultenabled>yes</defaultenabled>
 	<support_level>extended</support_level>
  ***/
 
--- a/addons/chan_ooh323.c
+++ b/addons/chan_ooh323.c
@@ -18,7 +18,7 @@
 
 
 /*** MODULEINFO
-	<defaultenabled>no</defaultenabled>
+	<defaultenabled>yes</defaultenabled>
 	<support_level>extended</support_level>
  ***/
 
--- a/addons/res_config_mysql.c
+++ b/addons/res_config_mysql.c
@@ -24,7 +24,7 @@
 
 /*** MODULEINFO
 	<depend>mysqlclient</depend>
-	<defaultenabled>no</defaultenabled>
+	<defaultenabled>yes</defaultenabled>
 	<support_level>extended</support_level>
  ***/
 
--- a/res/res_pktccops.c
+++ b/res/res_pktccops.c
@@ -31,7 +31,7 @@
  */
 
 /*** MODULEINFO
-        <defaultenabled>no</defaultenabled>
+        <defaultenabled>yes</defaultenabled>
 	<support_level>extended</support_level>
  ***/
 
