--- a/configure
+++ b/configure
@@ -16784,7 +16784,7 @@ if ac_fn_c_try_link "$LINENO"; then :
   { $as_echo "$as_me:${as_lineno-$LINENO}: result: yes" >&5
 $as_echo "yes" >&6; }
 
-$as_echo "#define HAVE_RES_NINIT 1" >>confdefs.h
+$as_echo "//#define HAVE_RES_NINIT 1" >>confdefs.h
 
 	{ $as_echo "$as_me:${as_lineno-$LINENO}: checking for library containing res_9_ndestroy" >&5
 $as_echo_n "checking for library containing res_9_ndestroy... " >&6; }
