--- a/ppdc/Makefile
+++ b/ppdc/Makefile
@@ -239,12 +239,7 @@ framedhelp:
 
 genstrings:		genstrings.o libcupsppdc.a ../cups/libcups.a \
 			sample.drv ../data/media.defs
-	echo Linking $@...
-	$(CXX) $(ARCHFLAGS) $(LDFLAGS) -o genstrings genstrings.o \
-		libcupsppdc.a ../cups/libcups.a $(LIBGSSAPI) $(SSLLIBS) \
-		$(DNSSDLIBS) $(COMMONLIBS) $(LIBZ)
-	echo Generating localization strings...
-	./genstrings >sample.c
+	echo Not Linking $@...
 
 
 #
