---
 Makefile |   20 ++++++++++----------
 1 file changed, 10 insertions(+), 10 deletions(-)

--- st2205tool.orig/Makefile
+++ st2205tool/Makefile
@@ -6,27 +6,27 @@ LIBS	=
 all:	libst2205/libst2205.so setpic/setpic phack splice bgrep
 
 install: all
-	make -C libst2205 install
+	$(MAKE) -C libst2205 install
 
 libst2205/libst2205.so: 
-	make -C libst2205
+	$(MAKE) -C libst2205
 
-setpic/setpic:
-	make -C setpic
+setpic/setpic: libst2205/libst2205.so
+	$(MAKE) -C setpic
 
 phack:	$(OBJ) $(SRC)
-	gcc -o $(@) $(OBJ) $(LIBS) $(FLAGS)
+	$(GCC) -o $(@) $(OBJ) $(LIBS) $(FLAGS)
 
 splice:	splice.o splice.c
-	gcc -o splice splice.o
+	$(GCC) -o splice splice.o
 
 bgrep:	bgrep.o bgrep.c
-	gcc -o bgrep bgrep.o
+	$(GCC) -o bgrep bgrep.o
 
 clean:	
-	make -C libst2205 clean
-	make -C setpic clean
+	$(MAKE) -C libst2205 clean
+	$(MAKE) -C setpic clean
 	rm -f $(OBJ) phack splice splice.o bgrep bgrep.o
 
 distclean: clean
-	rm -f fwimage.bak memimage.bak fwimage.bin
\ No newline at end of file
+	rm -f fwimage.bak memimage.bak fwimage.bin
