--- st2205tool/setpic/Makefile.orig	2008-01-20 19:50:11.000000000 +0100
+++ st2205tool/setpic/Makefile	2008-04-22 16:06:40.000000000 +0200
@@ -1,12 +1,12 @@
 SRC	=	main.c
 OBJ	=	main.o
-CFLAGS	=	-g -Wall -I../libst2205
-LIBS	=	-lgd -L../libst2205 -lst2205 
+#CFLAGS	=	-g -Wall -I./../libst2205
+#LIBS	=	-lgd -lst2205 
 
 all:	setpic
 
 setpic:	$(OBJ) $(SRC)
-	gcc -o $(@) $(OBJ) $(LIBS) $(FLAGS)
+	$(GCC) -o $(@) $(OBJ) $(SETPICLIBS) $(CFLAGS)
 
 clean:	
 	rm -f $(OBJ) setpic
