--- a/Makefile.am
+++ b/Makefile.am
@@ -8,7 +8,6 @@ aclocal_DATA = gst-element-check-@GST_MA
 
 SUBDIRS = pkgconfig \
 	gst libs plugins tools tests \
-	docs \
 	po \
 	m4 \
 	common
@@ -16,7 +15,6 @@ SUBDIRS = pkgconfig \
 # These are all the possible subdirs
 DIST_SUBDIRS = pkgconfig \
 	gst libs plugins tools tests \
-	docs \
 	po \
 	m4 \
 	common
--- a/Makefile.in
+++ b/Makefile.in
@@ -445,7 +445,6 @@ aclocaldir = $(datadir)/aclocal
 aclocal_DATA = gst-element-check-@GST_MAJORMINOR@.m4
 SUBDIRS = pkgconfig \
 	gst libs plugins tools tests \
-	docs \
 	po \
 	m4 \
 	common
@@ -454,7 +453,6 @@ SUBDIRS = pkgconfig \
 # These are all the possible subdirs
 DIST_SUBDIRS = pkgconfig \
 	gst libs plugins tools tests \
-	docs \
 	po \
 	m4 \
 	common
