--- a/utils/Makefile
+++ b/utils/Makefile
@@ -5,21 +5,6 @@ all install: 
 		$(MAKE) -C $$i $@ || exit 1; \
 	done
 
-	# Test whether qmake is installed, and whether it is for qt4.
-	@if which qmake-qt4 >/dev/null 2>&1; then \
-		QMAKE=qmake-qt4; \
-	else \
-		QMAKE=qmake; \
-	fi; \
-	if which $$QMAKE >/dev/null 2>&1; then \
-		if $$QMAKE --version 2>&1 | grep '4\.[0-9][0-9]*\.[0-9][0-9]*' >/dev/null; then \
-			if [ ! -f qv4l2/Makefile ]; then \
-				(cd qv4l2 && $$QMAKE) || exit 1; \
-			fi; \
-			$(MAKE) -C qv4l2 -f Makefile.install $@; \
-		fi \
-	fi
-
 sync-with-kernel:
 	$(MAKE) -C keytable $@
 	$(MAKE) -C v4l2-dbg $@
