diff -ruN a/Makefile b/Makefile
--- a/Makefile	2009-05-07 14:31:14.000000000 +0100
+++ b/Makefile	2009-05-07 14:31:47.000000000 +0100
@@ -32,7 +32,7 @@
 # if your system lacks getopt_long, remove the comment from this line
 OBJS = redir.o $(GETOPT_OBJS)
 
-CFLAGS = -O2 -Wall $(STR_CFLAGS) $(WRAP_CFLAGS) $(EXTRA_CFLAGS)
+CFLAGS += -Wall $(STR_CFLAGS) $(WRAP_CFLAGS) $(EXTRA_CFLAGS)
 LDFLAGS = -s
 
 # solaris, and others, may also need these libraries to link
