--- a/configure
+++ b/configure
@@ -5459,7 +5459,7 @@ echo $ECHO_N "checking \"for lzo_asm.h\"
 	fi
         echo "$as_me:$LINENO: result: ($dir) yes " >&5
 echo "${ECHO_T}($dir) yes " >&6
-	LZOCHK="lzo2 lzo"
+	LZOCHK="lzo2"
     else
         echo "$as_me:$LINENO: result: \"no\"" >&5
 echo "${ECHO_T}\"no\"" >&6
@@ -5504,8 +5504,7 @@ echo "${ECHO_T}($dir) yes " >&6
         echo "$as_me:$LINENO: result: \"no\"" >&5
 echo "${ECHO_T}\"no\"" >&6
 	{ { echo "$as_me:$LINENO: error: LZO library not found. " >&5
-echo "$as_me: error: LZO library not found. " >&2;}
-   { (exit 1); exit 1; }; }
+echo "$as_me: error: LZO library not found. " >&2;} }
 
     fi
 
@@ -5598,13 +5597,11 @@ fi
       done
       if test $havelzo = 0; then
 	    { { echo "$as_me:$LINENO: error: LZO library not found. " >&5
-echo "$as_me: error: LZO library not found. " >&2;}
-   { (exit 1); exit 1; }; }
+echo "$as_me: error: LZO library not found. " >&2;} }
       fi
    else
         { { echo "$as_me:$LINENO: error: LZO headers not found. " >&5
-echo "$as_me: error: LZO headers not found. " >&2;}
-   { (exit 1); exit 1; }; }
+echo "$as_me: error: LZO headers not found. " >&2;} }
    fi
 fi
 
