Index: crtmpserver-r726/builders/make/compile.mk
===================================================================
--- crtmpserver-r726.orig/builders/make/compile.mk
+++ crtmpserver-r726/builders/make/compile.mk
@@ -43,10 +43,16 @@ FEATURES_DEFINES = \
 	-DHAS_PROTOCOL_RTP \
 	-DHAS_PROTOCOL_TS \
 	-DHAS_PROTOCOL_VAR \
+	-DHAS_PROTOCOL_CLI \
+	-DHAS_PROTOCOL_HLS \
+	-DHAS_PROTOCOL_RAWHTTPSTREAM \
 	-DHAS_LUA \
 	-DHAS_MEDIA_MP3 \
 	-DHAS_MEDIA_MP4 \
-	-DHAS_MEDIA_FLV
+	-DHAS_MEDIA_FLV \
+	-DHAS_MEDIA_NSV \
+	-DHAS_SYSLOG
+	
 
 
 DEFINES = $(PLATFORM_DEFINES) $(FEATURES_DEFINES)
