--- a/arch/mips/include/asm/thread_info.h
+++ b/arch/mips/include/asm/thread_info.h
@@ -87,6 +87,7 @@ register struct thread_info *__current_t
 
 #define STACK_WARN	(THREAD_SIZE / 8)
 
+#if 0
 #define __HAVE_ARCH_THREAD_INFO_ALLOCATOR
 
 #ifdef CONFIG_DEBUG_STACK_USAGE
@@ -98,6 +99,7 @@ register struct thread_info *__current_t
 #endif
 
 #define free_thread_info(info) kfree(info)
+#endif
 
 #define PREEMPT_ACTIVE		0x10000000
 
