Index: linux-3.4.25_openwrt/fs/yaffs2/yaffs_guts.c
===================================================================
--- linux-3.4.25_openwrt.orig/fs/yaffs2/yaffs_guts.c	2013-06-19 16:06:29.000000000 +0300
+++ linux-3.4.25_openwrt/fs/yaffs2/yaffs_guts.c	2013-06-23 15:28:46.000000000 +0300
@@ -1709,11 +1709,11 @@
 	}
 
 	/* TODO: Do we need this different handling for YAFFS2 and YAFFS1?? */
-	if (obj->my_dev->param.is_yaffs2)
+	// if (obj->my_dev->param.is_yaffs2) 
 		unlinkOp = (new_dir == obj->my_dev->unlinked_dir);
-	else
+	/* else
 		unlinkOp = (new_dir == obj->my_dev->unlinked_dir
-			    && obj->variant_type == YAFFS_OBJECT_TYPE_FILE);
+			    && obj->variant_type == YAFFS_OBJECT_TYPE_FILE); */
 
 	deleteOp = (new_dir == obj->my_dev->del_dir);
 
