diff -urN ettercap-NG-0.7.3/configure ettercap-NG-0.7.3.new/configure
--- ettercap-NG-0.7.3/configure	2005-05-27 17:11:45.000000000 +0200
+++ ettercap-NG-0.7.3.new/configure	2006-08-21 20:15:07.000000000 +0200
@@ -27354,6 +27354,7 @@
 
    echo "$as_me:$LINENO: checking for NS_GET32" >&5
 echo $ECHO_N "checking for NS_GET32... $ECHO_C" >&6
+cross_compiling=no
    if test "$cross_compiling" = yes; then
   { { echo "$as_me:$LINENO: error: cannot run test program while cross compiling
 See \`config.log' for more details." >&5
diff -urN ettercap-NG-0.7.3/configure.in ettercap-NG-0.7.3.new/configure.in
--- ettercap-NG-0.7.3/configure.in	2005-01-13 10:24:28.000000000 +0100
+++ ettercap-NG-0.7.3.new/configure.in	2006-08-21 20:13:29.000000000 +0200
@@ -236,7 +236,7 @@
 AC_CHECK_HEADERS(sys/poll.h poll.h sys/select.h)
 AC_CHECK_HEADERS(sys/cdefs.h)
 dnl nameser NS_GET32 and so on...
-EC_NS_GET()
+#EC_NS_GET()
 
 dnl ==================================================================
 dnl   Checks for typedefs, structures, and compiler characteristics.
