--- a/hidd/sdp.c
+++ b/hidd/sdp.c
@@ -34,6 +34,7 @@
 #include <sys/stat.h>
 #include <sys/param.h>
 #include <sys/socket.h>
+#include <linux/limits.h>
 
 #include <bluetooth/bluetooth.h>
 #include <bluetooth/l2cap.h>
--- a/tools/hciconfig.c
+++ b/tools/hciconfig.c
@@ -37,6 +37,7 @@
 #include <sys/param.h>
 #include <sys/ioctl.h>
 #include <sys/socket.h>
+#include <linux/limits.h>
 
 #include <bluetooth/bluetooth.h>
 #include <bluetooth/hci.h>
