--- a/input/device.c
+++ b/input/device.c
@@ -25,6 +25,8 @@
 #include <config.h>
 #endif
 
+#define ENOKEY          161     /* Required key not available */
+
 #include <stdlib.h>
 #include <errno.h>
 #include <fcntl.h>
--- a/input/storage.c
+++ b/input/storage.c
@@ -25,6 +25,8 @@
 #include <config.h>
 #endif
 
+#define ENOKEY          161     /* Required key not available */
+
 #include <stdio.h>
 #include <errno.h>
 #include <ctype.h>
