--- a/configure.ac
+++ b/configure.ac
@@ -227,7 +227,6 @@ ALL_LINGUAS="da de es eu fr hu it ja nl
 GP_GETTEXT_HACK([${PACKAGE}-${LIBGPHOTO2_CURRENT_MIN}],[The gPhoto Team],[${MAIL_GPHOTO_TRANSLATION}])
 AM_GNU_GETTEXT_VERSION([0.14.1])
 AM_GNU_GETTEXT([external])
-AM_PO_SUBDIRS()
 AM_ICONV()
 GP_GETTEXT_FLAGS()
 
--- a/libgphoto2_port/configure.ac
+++ b/libgphoto2_port/configure.ac
@@ -116,7 +116,6 @@ GP_GETTEXT_HACK([${PACKAGE}-${LIBGPHOTO2
 ALL_LINGUAS="cs da de fr es eu it ja nl pl pt_BR ru sk sv uk vi zh_CN zh_TW"
 AM_GNU_GETTEXT_VERSION([0.14.1])
 AM_GNU_GETTEXT([external])
-AM_PO_SUBDIRS()
 AM_ICONV()
 GP_GETTEXT_FLAGS()
 
--- a/libgphoto2_port/libgphoto2_port/Makefile.in
+++ b/libgphoto2_port/libgphoto2_port/Makefile.in
@@ -15,6 +15,8 @@
 
 @SET_MAKE@
 
+SHELL = @SHELL@
+
 VPATH = @srcdir@
 pkgdatadir = $(datadir)/@PACKAGE@
 pkgincludedir = $(includedir)/@PACKAGE@
--- a/libgphoto2_port/Makefile.in
+++ b/libgphoto2_port/Makefile.in
@@ -15,7 +15,7 @@
 
 @SET_MAKE@
 
-
+SHELL = @SHELL@
 
 VPATH = @srcdir@
 pkgdatadir = $(datadir)/@PACKAGE@
--- a/Makefile.in
+++ b/Makefile.in
@@ -20,7 +20,7 @@
 ########################################################################
 # put file lists and special compile rules here
 
-
+SHELL = @SHELL@
 
 VPATH = @srcdir@
 pkgdatadir = $(datadir)/@PACKAGE@
