--- a/Makefile.in
+++ b/Makefile.in
@@ -147,7 +147,7 @@ AM_RECURSIVE_TARGETS = $(RECURSIVE_TARGE
 	distdir dist dist-all distcheck
 ETAGS = etags
 CTAGS = ctags
-DIST_SUBDIRS = lib libextra gl src doc tests guile
+DIST_SUBDIRS = lib libextra gl src guile
 DISTFILES = $(DIST_COMMON) $(DIST_SOURCES) $(TEXINFOS) $(EXTRA_DIST)
 distdir = $(PACKAGE)-$(VERSION)
 top_distdir = $(distdir)
--- a/Makefile.am
+++ b/Makefile.am
@@ -22,7 +22,7 @@
 
 DISTCHECK_CONFIGURE_FLAGS = --enable-gtk-doc
 
-SUBDIRS = lib libextra gl src doc tests
+SUBDIRS = lib libextra gl src
 
 if HAVE_GUILE
 SUBDIRS += guile
--- a/lib/Makefile.in
+++ b/lib/Makefile.in
@@ -274,7 +274,7 @@ AM_RECURSIVE_TARGETS = $(RECURSIVE_TARGE
 	distdir dist dist-all distcheck
 ETAGS = etags
 CTAGS = ctags
-DIST_SUBDIRS = gl po includes x509 minitasn1 opencdk openpgp
+DIST_SUBDIRS = gl includes x509 minitasn1 opencdk openpgp
 DISTFILES = $(DIST_COMMON) $(DIST_SOURCES) $(TEXINFOS) $(EXTRA_DIST)
 distdir = $(PACKAGE)-$(VERSION)
 top_distdir = $(distdir)
@@ -819,7 +819,7 @@ top_build_prefix = @top_build_prefix@
 top_builddir = @top_builddir@
 top_srcdir = @top_srcdir@
 ACLOCAL_AMFLAGS = -I m4 -I gl/m4
-SUBDIRS = gl po includes x509 $(am__append_1) $(am__append_2)
+SUBDIRS = gl includes x509 $(am__append_1) $(am__append_2)
 AM_CFLAGS = $(WARN_CFLAGS) $(WERROR_CFLAGS)
 AM_CPPFLAGS = -DLOCALEDIR=\"$(localedir)\" -I$(srcdir)/gl \
 	-I$(builddir)/gl -I$(srcdir)/includes -I$(builddir)/includes \
--- a/lib/Makefile.am
+++ b/lib/Makefile.am
@@ -22,7 +22,7 @@
 
 ACLOCAL_AMFLAGS = -I m4 -I gl/m4
 
-SUBDIRS = gl po includes x509
+SUBDIRS = gl includes x509
 if ENABLE_MINITASN1
 SUBDIRS += minitasn1
 endif
--- a/lib/configure.ac
+++ b/lib/configure.ac
@@ -37,9 +37,6 @@ AC_PROG_LIBTOOL
 
 LIBGNUTLS_HOOKS
 
-AM_GNU_GETTEXT([external])
-AM_GNU_GETTEXT_VERSION([0.17])
-
 AC_C_BIGENDIAN
 
 # For includes/gnutls/gnutls.h.in.
