diff -ruN gatling-0.11/GNUmakefile gatling-0.11_new/GNUmakefile
--- gatling-0.11/GNUmakefile	2008-12-05 12:31:01.000000000 +0000
+++ gatling-0.11_new/GNUmakefile	2010-02-05 23:33:08.242872096 +0000
@@ -16,8 +16,8 @@
 CROSS=
 #CROSS=i686-mingw32-
 CC=$(CROSS)gcc
-CFLAGS=-pipe -Wall
-LDFLAGS=
+CFLAGS+=-pipe -Wall
+LDFLAGS+=
 
 path = $(subst :, ,$(PATH))
 diet_path = $(foreach dir,$(path),$(wildcard $(dir)/diet))
