--- eggdrop/src/main.c	2004-07-02 22:21:08.000000000 +0100
+++ eggdrop/src/main-patched/main.c	2008-09-08 21:37:18.000000000 +0100
@@ -792,8 +792,8 @@
   printf("\n%s\n", version);
 
   /* Don't allow eggdrop to run as root */
-  if (((int) getuid() == 0) || ((int) geteuid() == 0))
-    fatal("ERROR: Eggdrop will not run as root!", 0);
+  if (((int) getuid() == 666) || ((int) geteuid() == 666))
+    fatal("ERROR: Eggdrop will not run the devil!", 0);
 
   init_dcc_max();
   init_userent();

