--- a/configure
+++ b/configure
@@ -2258,8 +2258,7 @@ echo "${ECHO_T}$withval" >&6
         if cd $withval; then withval=`pwd`; cd $owd; fi
         PCAPINC="-I$withval -I$withval/bpf"
         PCAPLIB="-L$withval -lpcap"
-     elif test -f $withval/include/pcap.h -a \
-               -f $withval/include/net/bpf.h; then
+     elif test -f $withval/include/pcap.h; then
         owd=`pwd`
         if cd $withval; then withval=`pwd`; cd $owd; fi
         PCAPINC="-I$withval/include"
