--- a/build/modules.conf.in
+++ b/build/modules.conf.in
@@ -41,10 +41,14 @@ applications/mod_valet_parking
 #applications/mod_ladspa
 #applications/mod_mongo
 applications/mod_sms
+#applications/mod_http_cache
+#applications/mod_mp4
+#applications/mod_protovm
 codecs/mod_g723_1
 codecs/mod_amr
 #codecs/mod_amrwb
 #codecs/mod_silk
+#codecs/mod_theora
 #codecs/mod_codec2
 codecs/mod_g729
 #codecs/mod_com_g729
@@ -62,6 +66,7 @@ dialplans/mod_dialplan_xml
 dialplans/mod_dialplan_asterisk
 #directories/mod_ldap
 #endpoints/mod_dingaling
+#endpoints/mod_gsmopen
 #endpoints/mod_portaudio
 endpoints/mod_sofia
 endpoints/mod_loopback
@@ -82,6 +87,7 @@ endpoints/mod_loopback
 #event_handlers/mod_event_multicast
 event_handlers/mod_event_socket
 #event_handlers/mod_event_zmq
+#event_handlers/mod_json_cdr
 event_handlers/mod_cdr_csv
 event_handlers/mod_cdr_sqlite
 #event_handlers/mod_cdr_mongodb
