Index: freeswitch-1.0.7/libs/spandsp/src/spandsp/fast_convert.h
===================================================================
--- freeswitch-1.0.7.orig/libs/spandsp/src/spandsp/fast_convert.h	2012-08-14 19:52:20.000000000 +0200
+++ freeswitch-1.0.7/libs/spandsp/src/spandsp/fast_convert.h	2012-08-14 19:53:01.000000000 +0200
@@ -195,7 +195,7 @@
     {
         return (long int) (x);
     }
-#elif defined(__ppc__)  ||   defined(__powerpc__)
+#elif 0 /* defined(__ppc__)  ||   defined(__powerpc__) */
     static __inline__ long int lfastrint(register double x)
     {
         int res[2];
