--- a/cgitest.c
+++ b/cgitest.c
@@ -216,7 +216,7 @@ int main (int argc, char **argv, char **
 	    cgiSetHeader ("Set-Cookie", "Version=1; Library=cgilib; Path=/");
             cgiHeader();
 	    printf (HTML_CSS HTML_HEAD);
-	    printf ("<h3>Cookie Library set</h3>\n");
+	    printf ("<h3>Cookie \"Library\" set</h3>\n");
 	    printf ("<p><br><p><br><a href=\"/cgi-bin/cgitest\">Test</a><p>\n");
 	    printf ("<p><br><p><br><a href=\"/cgi-bin/cgitest/redirect\">Redirect</a><p>\n");
 	    printf ("<p><br><p><br><a href=\"/cgi-bin/cgitest/listall\">List Everything</a><p>\n");
