diff --git a/plugins/video/common/trace.cxx b/plugins/video/common/trace.cxx
index 7262478..de641af 100644
--- a/plugins/video/common/trace.cxx
+++ b/plugins/video/common/trace.cxx
@@ -30,6 +30,7 @@
 /*****************************************************************************/
 
 #include "trace.h"
+#include <cstring>
 
 #ifdef __MACOSX__
 #include <libgen.h>
diff --git a/plugins/video/H.263-ffmpeg/h263ffmpeg.cxx b/plugins/video/H.263-ffmpeg/h263ffmpeg.cxx
index 4a17b2c..a2f051c 100644
--- a/plugins/video/H.263-ffmpeg/h263ffmpeg.cxx
+++ b/plugins/video/H.263-ffmpeg/h263ffmpeg.cxx
@@ -54,6 +54,8 @@
 #define _CRT_NONSTDC_NO_WARNINGS
 #define _CRT_SECURE_NO_DEPRECATE
 #include <codec/opalplugin.h>
+#include <cstdio>
+#include <climits>
 
 #include <stdlib.h>
 #ifdef _WIN32
