--- a/config.h
+++ b/config.h
@@ -48,7 +48,7 @@
 
 #define MAX_LISTEN 16		/*!< maximum number of addresses on which we will listen */
 
-#define CHILD_NO    8		/*!< default number of child processes started */
+#define CHILD_NO    4		/*!< default number of child processes started */
 
 #define RT_NO 100 		/*!< routing tables number */
 #define FAILURE_RT_NO RT_NO	/*!< on_failure routing tables number */
@@ -114,7 +114,7 @@
 #define SRV_MAX_PREFIX_LEN SRV_TLS_PREFIX_LEN
 
 #define PKG_MEM_POOL_SIZE 1024*1024		/*!< Used only if PKG_MALLOC is defined*/
-#define SHM_MEM_SIZE 32				/*!< Used if SH_MEM is defined*/
+#define SHM_MEM_SIZE 8				/*!< Used if SH_MEM is defined*/
 
 #define TIMER_TICK   1  			/*!< one second */
 #define UTIMER_TICK  100*1000			/*!< 100 miliseconds*/
