--- a/config.h
+++ b/config.h
@@ -48,7 +48,7 @@
 
 #define MAX_LISTEN 16		/*!< maximum number of addresses on which we will listen */
 
-#define CHILD_NO    8		/*!< default number of child processes started */
+#define CHILD_NO    4		/*!< default number of child processes started */
 
 #ifndef RT_NO
 #define RT_NO 80 		/*!< routing tables number */
@@ -119,7 +119,7 @@
 #define PKG_MEM_POOL_SIZE 4*1024*1024		/*!< Used only if PKG_MALLOC is defined*/
 #endif
 
-#define SHM_MEM_SIZE 32				/*!< Used if SH_MEM is defined*/
+#define SHM_MEM_SIZE 8				/*!< Used if SH_MEM is defined*/
 
 #define TIMER_TICK   1  			/*!< one second */
 #define UTIMER_TICK  100*1000			/*!< 100 miliseconds*/
