diff --git a/bin/flow-log2rrd b/bin/flow-log2rrd
index 0567ba5..db4fdda 100755
--- a/bin/flow-log2rrd
+++ b/bin/flow-log2rrd
@@ -1,4 +1,4 @@
-#!/usr/local/bin/python
+#!/bin/env python
 
 import getopt
 import os
diff --git a/bin/flow-rpt2rrd b/bin/flow-rpt2rrd
index 72a74c7..7698f2b 100755
--- a/bin/flow-rpt2rrd
+++ b/bin/flow-rpt2rrd
@@ -1,4 +1,4 @@
-#!/usr/local/bin/python
+#!/bin/env python
 
 import getopt
 import os
diff --git a/bin/flow-rptfmt b/bin/flow-rptfmt
index 047d1ba..f663fae 100755
--- a/bin/flow-rptfmt
+++ b/bin/flow-rptfmt
@@ -1,4 +1,4 @@
-#!/usr/local/bin/python
+#!/bin/env python
 
 import sys
 import string
