--- a/cgatool/cgatool.c
+++ b/cgatool/cgatool.c
@@ -49,7 +49,7 @@
 
 #include "config.h"
 #include <applog.h>
-#include <libconfig.h>
+#include "../libs/libconfig/libconfig.h"
 #include <senddctl.h>
 
 #include <cga.h>
--- a/sendd/config.c
+++ b/sendd/config.c
@@ -38,7 +38,7 @@
 #include <net/if.h>
 
 #include "config.h"
-#include <libconfig.h>
+#include "../libs/libconfig/libconfig.h"
 #include <applog.h>
 
 #include "sendd_local.h"
