--- a/Makefile.am
+++ b/Makefile.am
@@ -1,7 +1,7 @@
 
 AUTOMAKE_OPTIONS = foreign
 
-SUBDIRS = liboil testsuite examples doc
+SUBDIRS = liboil
 
 EXTRA_DIST = COPYING autogen.sh gtk-doc.make HACKING BUG-REPORTING
 
--- a/Makefile.in
+++ b/Makefile.in
@@ -278,7 +278,7 @@ top_build_prefix = @top_build_prefix@
 top_builddir = @top_builddir@
 top_srcdir = @top_srcdir@
 AUTOMAKE_OPTIONS = foreign
-SUBDIRS = liboil testsuite examples doc
+SUBDIRS = liboil
 EXTRA_DIST = COPYING autogen.sh gtk-doc.make HACKING BUG-REPORTING
 DISTCHECK_CONFIGURE_FLAGS = --enable-gtk-doc
 pkgconfig_DATA = liboil-$(LIBOIL_MAJORMINOR).pc
