--- a/src/chan_sccp.h
+++ b/src/chan_sccp.h
@@ -127,15 +127,7 @@ extern "C" {
 #        define CHECK_LEAKS()
 #    endif
 
-#    define SCCP_FILE_VERSION(file, version) \
-	static void __attribute__((constructor)) __register_file_version(void) \
-	{ \
-		pbx_register_file_version(file, version); \
-	} \
-	static void __attribute__((destructor)) __unregister_file_version(void) \
-	{ \
-		pbx_unregister_file_version(file); \
-	}
+#    define SCCP_FILE_VERSION(file, version) 
 
 #    define DEV_ID_LOG(x) x ? x->id : "SCCP"
 
