diff -urN urbi-sdk/src/Makefile.am urbi-sdk.new/src/Makefile.am
--- urbi-sdk/src/Makefile.am	2007-01-31 11:37:04.000000000 +0100
+++ urbi-sdk.new/src/Makefile.am	2007-10-03 11:17:00.000000000 +0200
@@ -1,5 +1,5 @@
 SUBDIRS = liburbi 
 
 if EXAMPLES
-SUBDIRS += utils examples
+SUBDIRS += utils
 endif
diff -urN urbi-sdk/src/Makefile.in urbi-sdk.new/src/Makefile.in
--- urbi-sdk/src/Makefile.in	2007-04-10 17:44:07.000000000 +0200
+++ urbi-sdk.new/src/Makefile.in	2007-10-03 11:14:33.000000000 +0200
@@ -30,7 +30,7 @@
 POST_UNINSTALL = :
 build_triplet = @build@
 host_triplet = @host@
-@EXAMPLES_TRUE@am__append_1 = utils examples
+@EXAMPLES_TRUE@am__append_1 = utils
 subdir = src
 DIST_COMMON = $(srcdir)/Makefile.am $(srcdir)/Makefile.in
 ACLOCAL_M4 = $(top_srcdir)/aclocal.m4
@@ -68,7 +68,7 @@
   distclean-recursive maintainer-clean-recursive
 ETAGS = etags
 CTAGS = ctags
-DIST_SUBDIRS = liburbi utils examples
+DIST_SUBDIRS = liburbi utils
 DISTFILES = $(DIST_COMMON) $(DIST_SOURCES) $(TEXINFOS) $(EXTRA_DIST)
 INSTALL_HEADER = @INSTALL_HEADER@
 ACLOCAL = @ACLOCAL@
