Index: veth-1.0/veth/Makefile
===================================================================
--- veth-1.0.orig/veth/Makefile	2008-10-29 16:11:33.000000000 +0100
+++ veth-1.0/veth/Makefile	2008-10-31 18:01:33.000000000 +0100
@@ -4,8 +4,7 @@
 	rm -f vethd
 
 vethd: vethd.c
-	$(CC) -O6 -Wall -pedantic -o vethd vethd.c
-	strip vethd
+	$(CC) $(CFLAGS) -o vethd vethd.c
 
 install: all
 	install -s -m 750 ./vethd $(INSTALL_DIR)
