diff -urN ndyndns-2.1.orig/configure ndyndns-2.1/configure
--- ndyndns-2.1.orig/configure	2010-07-17 13:22:39.000000000 +0200
+++ ndyndns-2.1/configure	2011-01-26 16:24:15.834445729 +0100
@@ -3318,7 +3318,7 @@
 fi
 
 
-CURLINC=-I`curl-config --prefix`/include
+CURLINC=-I`curl-config --cflags`
 
 CURLLIB=`curl-config --libs`
 
diff -urN ndyndns-2.1.orig/configure.in ndyndns-2.1/configure.in
--- ndyndns-2.1.orig/configure.in	2010-07-17 13:22:11.000000000 +0200
+++ ndyndns-2.1/configure.in	2011-01-26 16:24:14.090441650 +0100
@@ -24,7 +24,7 @@
 fi
 AC_SUBST(PLATFORM)
 
-CURLINC=-I`curl-config --prefix`/include
+CURLINC=-I`curl-config --cflags`
 AC_SUBST(CURLINC)
 CURLLIB=`curl-config --libs`
 AC_SUBST(CURLLIB)
