diff --git a/configure b/configure
index 37b66f3..723ca1e 100755
--- a/configure
+++ b/configure
@@ -10296,7 +10296,6 @@ done
 
 for ac_header in \
   string                        \
-  iomanip.h                     \
 
 do
 as_ac_Header=`echo "ac_cv_header_$ac_header" | $as_tr_sh`
diff --git a/configure.in b/configure.in
index c857735..abfe4cd 100644
--- a/configure.in
+++ b/configure.in
@@ -227,7 +227,6 @@ AC_CHECK_HEADERS(fstream iostream iomanip vector \
 )
 AC_CHECK_HEADERS(               \
   string                        \
-  iomanip.h                     \
   ,,AC_MSG_ERROR([Missing a vital header file for id3lib])
 )
 
