diff --git a/include/id3/id3lib_strings.h b/include/id3/id3lib_strings.h
index 02a9aba..101b797 100644
--- a/include/id3/id3lib_strings.h
+++ b/include/id3/id3lib_strings.h
@@ -29,6 +29,7 @@
 #ifndef _ID3LIB_STRINGS_H_
 #define _ID3LIB_STRINGS_H_
 
+#include <string.h>
 #include <string>
 #include <streambuf>
 
diff --git a/include/id3/writers.h b/include/id3/writers.h
index 1cbb0dc..4bc8f77 100644
--- a/include/id3/writers.h
+++ b/include/id3/writers.h
@@ -30,7 +30,7 @@
 
 #include "id3/writer.h"
 #include "id3/id3lib_streams.h"
-//#include <string.h>
+#include <string.h>
 
 class ID3_CPP_EXPORT ID3_OStreamWriter : public ID3_Writer
 {
