diff --git a/Makefile.in b/Makefile.in
index 8c121c5..b2131c2 100644
--- a/Makefile.in
+++ b/Makefile.in
@@ -132,8 +132,8 @@ EXTRA_DIST = \
 @ID3_NEEDZLIB_TRUE@zlib_subdir = zlib
 @ID3_NEEDZLIB_FALSE@zlib_subdir = 
 
-SUBDIRS = . m4 $(zlib_subdir) doc include id3com src examples
-DIST_SUBDIRS = . m4 zlib doc include id3com src examples prj libprj
+SUBDIRS = . m4 $(zlib_subdir) doc include id3com src
+DIST_SUBDIRS = . m4 zlib doc include id3com src prj libprj
 
 INCLUDES = @ID3LIB_DEBUG_FLAGS@
 
