--- a/core/encr/tkip.c
+++ b/core/encr/tkip.c
@@ -241,7 +241,7 @@ void generateRand(UINT8 *Data, UINT32 le
 
 	for (i = length;i--;)
 	{
-		Data[i] = net_random();
+		Data[i] = prandom_u32();
 	}
 }
 
--- a/core/mgt/AP/mlmeAuth.c
+++ b/core/mgt/AP/mlmeAuth.c
@@ -304,7 +304,7 @@ SINT32 mlmeAuthDoSharedKeySeq1(vmacApInf
 #ifdef ECOS
 			value = rand();
 #else
-			value = net_random(); //rand();
+			value = prandom_u32(); //rand();
 #endif
 			memcpy(&MgmtRsp->Body.Auth.ChallengeText.Text[i], &value, 4);
 			memcpy(&Challenge[ChallengeIdx].Text[i], &value, 4);
--- a/core/mgt/STA/mlmeApiSta.c
+++ b/core/mgt/STA/mlmeApiSta.c
@@ -3093,7 +3093,7 @@ extern UINT32 mlmeApiGetTime(vmacStaInfo
 extern UINT32 mlmeApiGenRandomUINT32(vmacStaInfo_t *vStaInfo_p, UINT32 seed)
 {
 #ifdef PORT_TO_LINUX_OS
-	return net_random();
+	return prandom_u32();
 #else /* PORT_TO_LINUX_OS */
 	srand(seed);
 	return rand();
