Index: rtorrent-0.8.6/scripts/common.m4
===================================================================
--- rtorrent-0.8.6.orig/scripts/common.m4
+++ rtorrent-0.8.6/scripts/common.m4
@@ -184,7 +184,7 @@ AC_DEFUN([TORRENT_CHECK_MADVISE], [
 AC_DEFUN([TORRENT_CHECK_EXECINFO], [
   AC_MSG_CHECKING(for execinfo.h)
 
-  AC_RUN_IFELSE(
+  AC_LINK_IFELSE(
     [[#include <execinfo.h>
       int main() { backtrace((void**)0, 0); backtrace_symbols((char**)0, 0); return 0;}
     ]],
@@ -199,7 +199,7 @@ AC_DEFUN([TORRENT_CHECK_EXECINFO], [
 AC_DEFUN([TORRENT_CHECK_ALIGNED], [
   AC_MSG_CHECKING(the byte alignment)
 
-  AC_RUN_IFELSE(
+  AC_LINK_IFELSE(
     [[#include <inttypes.h>
       int main() {
         char buf[8] = { 0, 0, 0, 0, 1, 0, 0, 0 };
