diff -urN patch-2.5.4.orig/Makefile.in patch-2.5.4/Makefile.in
--- patch-2.5.4.orig/Makefile.in	2008-06-11 10:08:08.000000000 +0200
+++ patch-2.5.4/Makefile.in	2008-06-11 10:08:26.000000000 +0200
@@ -42,12 +42,12 @@
 VERSION = @VERSION@
 
 prefix = @prefix@
-exec_prefix = @exec_prefix@
+exec_prefix += @exec_prefix@
 
 bindir = $(exec_prefix)/bin
 
 # Where to put the manual pages.
-mandir = @mandir@
+mandir = $(exec_prefix)/@mandir@
 man1dir = $(mandir)/man1
 # Extension (including `.') for the manual page filenames.
 man1ext = .1
