diff -Nurb a/src/Linux/Makefile b/src/Linux/Makefile
--- a/src/Linux/Makefile	2011-09-07 01:15:22.000000000 +0200
+++ b/src/Linux/Makefile	2012-03-21 01:55:40.468569332 +0100
@@ -97,7 +97,7 @@
 
 SFLOWOVS_OBJS=sflowovsd.o util.o
 
-CFLAGS+= -I. -I$(SFLOWDIR) $(OPT) -Wall -D_GNU_SOURCE -DHSP_VERSION=$(VERSION) -DUTHEAP
+CFLAGS+= -I. -I$(SFLOWDIR) $(OPT) -Wall -D_GNU_SOURCE -DHSP_VERSION=$(VERSION)
 LIBS+= $(SFLOWDIR)/libsflow.a -lresolv -lpthread
 
 #### BUILD ####
